![logo](https://github.com/shooshtarilab/RegSCOUT/blob/main/Logo/RegSCOUT_logo.png)

# RegSCOUT (<ins>Reg</ins>ulatory <ins>S</ins>ingle <ins>C</ins>ell <ins>O</ins>mics for <ins>U</ins>nravelling <ins>T</ins>rait Loci - Updates in Progress)
This is the repository for the RegSCOUT pipeline. This pipeline can be used as a stand-alone command-line application which integrates disease-associated SNPs, identified through genome-wide association studies (GWAS), with scATAC-seq, Hi-C, and eQTL data. Through this integrative analysis RegSCOUT prioritizes putative gene regulatory networks relevant to complex disease pathogenesis. This pipeline can handle both a scATAC-seq data as a Seurat object or processed results of analyzing scATAC-seq data as a peak and cell type table along with peak-peak interaction tables (generated by [cicero](https://bioconductor.org/packages/release/bioc/html/cicero.html)). See the --mode peak_table section for more information about these two tables. Example bash scripts for use of RegSCOUT can be found in the bash_scripts folder on this repository. 

## Updates as of July 20, 2025
Hi-C, eQTL, TF expression, histone mark analysis, as well as gene prioritization codes have been completed. Example bash scripts have been updated to include these changes. Bash scripts have not yet been tested, I apologize for any bugs in those scripts, however, the analyses conducted in these new R scripts have been tested on my local machine. Please don't hesitate to let me know if any urgent changes are required to any aspect of the pipeline. I recommend re-downloading all files off the github before running RegSCOUT as minor changes may have occurred in earlier stages of the pipeline e.g., fgwas_data_prep.R to create files easier to work with in later stages of the pipeline e.g., final_output.R. Each of the new steps: Hi-C, eQTL, TF expression, and histone mark analysis (but not gene prioritization (final_output.R)) will require an excel spreadsheet be made and directory provided, see below for more info. Example spreadsheets can be found in the instructions_spreadsheets folder on this github.

### Default parameters
Default parameters have been added to most aspects of the pipeline where they could help the user save writing out a parameter in the bash script. The example bash scripts have all parameters written but for parameters like LD_thr or cic_genomic_window there are defaults and the user should not need to provide values for the pipeline to still be able to run (this is untested). If in doubt about which parameters now have a default (this info will be added to the github in the near future), see first couple lines of R script for list called defaults (if this list is not present, there are no defaults for that stage of RegSCOUT). 

### New parameters in earlier stages of pipeline
*LD_thr* -> in fgwas_data_prep: allows user to set the threshold for LD filtering, default is 0.25
*CI_thr* -> not a new parameter, but there was an error in earlier versions of the bash script where it was set to 95 (please change this to 0.95 or a different cutoff if desired), the default is 0.95.
*locus_region* -> in fgwas_data_prep: defines the size of the locus region, setting this to 1000000 (meaning 1Mbp upstream/downstream a lead SNP) defines a 2 Mbp region around each lead SNP. Default value is 1000000 (1Mbp).
*cic_genomic_window* -> in peak_interaction_extract: defines the maximum distance apart two open chromatin regions can be for cicero to calculate a co-accessibility score. Default is 2000000 (2Mbp).
*coaccess_th* -> in peak_interaction_extract: the cutoff threshold for the co-accessibility value between two peaks, for that peak-peak association to be output in this step. Default is 0.05.

### Hi-C and eQTL analysis
If these are desired, set the hic_eqtl_analysis equal to Y. 

Hi-C analysis required parameters: output_dir, hic_instruct_dir, prom_th_up, prom_th_down, and gencode_dir. The parameter hic_instruct_dir should be set to the *path* to the spreadsheet for Hi-C analysis. The spreadsheet for Hi-C analysis contains 6 columns: hic_dir (directory of each Hi-C dataset), hic_type (either intact or PCHi-C), bulk (TRUE or FALSE), atac_cell_types (comma but not space separated cell types whose analysis should be conducted with that specific Hi-C dataset), hic_cell_types (the cell types in the Hi-C dataset that should be used to analyze each of the atac cell types, note how the order of Hi-C cell types should match the order of atac cell types, also note that the same cell type can repeat twice if used to analyze more than one atac cell type), and signif_threshold (the threshold for significance filtering only used if the Hi-C dataset provided contains more than one cell type). 

To be more specific. RegSCOUT will work no matter if it is provided a Hi-C dataset with more than one cell type, or if it is provided with Hi-C datasets that only contain one cell type each, or a mix. RegSCOUT expects all Hi-C interactions to have already been filtered for significance before being provided to the pipeline, if the Hi-C results are bulk or only relevant to one cell type. Necessary columns in this case for intact are: Chr1, Start1, End1, Chr2, Start2, and End2; for bulk: baitChr, baitStart, baitEnd, baitName, oeChr, oeStart, and oeEnd. In the case where Hi-C results are bulk please indicate that in the spreadsheet and in the case that the Hi-C is only relevant to one cell type, please leave the hic_cell_types column blank. If a Hi-C dataset is relevant to more than just one cell type, fill in the hic_cell_types column and signif threshold. Each cell type should now have its own column, in these columns should be the significance scores. The names of these cell type columns should match those provided in hic_cell_types. 

Output of Hi-C analysis will include rmpRegion (region of RMP), oeRegion (Hi-C region that overlapped RMP), promRegion (Hi-C region that overlapped promoter), cell, and gene columns.

eQTL analysis required parameters: output_dir, and eqtl_instruct_dir. The parameter eqtl_instruct_dir should provide the path to the instructions spreadsheet. This spreadsheet only has two columns: eqtl_dir (the path to each eQTL dataset) and atac_cell_types (which cell types each eQTL dataset will be used to conduct analysis for). RegSCOUT does not filter eQTLs for significance, the user must ensure this has been done prior to providing that to the RegSCOUT. Note multiple atac cell types can be associated with an eQTL dataset, this should only happen if that eQTL dataset is for example representative of T cells while the atac cell types are subtypes of T cells. RegSCOUT will assume the eQTL dataset is relevant to each of the atac cell types if multiple are listed, it will use the entire eQTL dataset provided to investigate SNPs in each of those cell types (it won't partition the eQTL dataset into eQTLs specific to one T cell subtype and eQTLs specific to the other for example).

### TF expression analysis
This all relates to the tf_expr_analysis parameter which can be set to one of four different values: none, atac, rna, or both. If set to both, RegSCOUT will conduct TF expression analysis using both scATAC-seq (using the peaks found in peak_cell_extract.R) and scRNA-seq (using seurat objects that the user must provide). If it is set to none, no analysis will be conducted regarding TF expression, and if set to one of atac or rna, it will conduct the requested analysis. 

Required parameters: output_dir, prom_th_up, prom_th_down, gencode_dir, tf_rna_quantile_th (if asking for scRNA-seq analysis, the threshold at which TFs are considered to not be expressed in a cell, default is 0.25), scrna_instruct_dir (if asking for scRNA-seq analysis). 

The parameter scrna_instruct_dir indicates the path to the scRNA-seq analysis spreadsheet. Five columns are present in this spreadsheet: scrna_dir (path to each scRNA-seq dataset), atac_cell_types (the atac cell types that each scRNA-seq dataset will be used to analyze), rna_cell_types (the cell types in the scRNA-seq dataset that will be used for scRNA-seq analysis, order of these cell types must match order of atac cell types, similar to what is required for Hi-C analysis), one_cell_type_seurat (RegSCOUT will work if a seurat object with many cell types is provided, if a seurat with just one is provided, or if a mix of those seurat objects are provided, set this to TRUE if the seurat object has a gene by cell matrix for one cell type), and matrix_loc (the location of the gene by cell matrix, either counts or data). Like with Hi-C analysis, rna_cell_type naming must match exactly the order of the atac cell types, thus rna cell type names can repeat if it is used to confirm more than one atac cell type. If a seurat is a one cell type seurat, it is optional to list rna cell types, that column is not used for analysis of one cell type seurats. 

Output of this step is a TF by cell type matrix. 'atac' means a TF was supported by scATAC-seq data, 'rna' means a TF was supported scRNA-seq data, and 'both' means a TF was supported by both data modalities.

### Histone mark analysis 
Like the previous analyses, this analysis assumes the files provided by the user have already been filtered for statistical significance. Required parameters: output_dir and hist_mark_instruct_dir (path to histone mark instructions spreadsheet). The histone mark instructions spreadsheet should contain three columns: chromHMM_dir (path to chromHMM results), atac_cell_types (those cell types in ATAC-seq whose peaks are to be confirmed by a specific chromHMM dataset), and chromHMM_cell_types (these cell type names can be anything defined by the user for example a dataset encompassing dendritic cells could be named dc_his, there should be only one cell type listed in each cell of this column). 

### Creating final tables
Required parameters: output_dir and gene_ppa_th (the gene ppa cutoff above which a gene is prioritized (assuming other criteria are met i.e., it must be supported by at least 2 methods, etc.; the default value is 0.05)).

There are three outputs of this code: final_table.xlsx (no TF expression information, no gene prioritization has been done, this shows all eQTL, Hi-C, cicero, and direct overlap genes among many other columns), complete_table.xlsx (TF expression information included, no gene prioritization has been done, gene ppas and gene scores are listed), and prioritized_table (gene prioritization has been conducted (no gene should have gene ppa <= 0.05 or gene score < 2), if Hi-C and eQTL analyses were conducted). 

## Required R version and libraries:
R version 4.4.0 or later and these libraries: Seurat, Signac, GenomicRanges, genetics.binaRies, R.utils, atSNP, TFBSTools, stringr, circlize, ComplexHeatmap, xlsx, R.utils, ape, monocle3, cicero, Repitools, reshape2, dplyr, BSgenome.Hsapiens.UCSC.hg19, BSgenome.Hsapiens.UCSC.hg38, ggplot2, readxl, tidyr, tibble. It is recommended that monocle3 be installed prior to installing cicero. Cicero should be installed following instructions found in this link: https://cole-trapnell-lab.github.io/cicero-release/docs_m3/. 

Descriptions of required parameters and of RegSCOUT output can be found below. 

## Main Parameters:
### --mode: 
This parameter should be used to choose the mode of operation of the pipeline based on input type. If a scATAC-seq Seurat object is provided, this option should be set to ATAC_obj (i.e., --mode ATAC_obj). If a peak and cell type table and peak-peak interaction table are used this mode should be set to peak_table (i.e., --mode peak_table).

### --finemap:
This parameter should be used only when the user wants fine-mapping to be conducted within the RegSCOUT workflow using [fgwas](https://github.com/joepickrell/fgwas). If this is the case --finemap should be set to Y (i.e., --finemap Y). If this is not the case, this parameter should not be used. 

### --output_dir: 
This parameter should be used for setting the working directory (the directory address should end with "/"). **This parameter should always be set when using RegSCOUT regardless of the value of --finemap or --mode**. All the output and intermediate files of the pipeline are generated in this directory. If running the pipeline with --mode peak_table, the peak-by-cell and peak-peak interaction tables of the pipeline should be provided in this directory.

*All directories specified in parameters should end with "/"*

*Please note that case (lower/upper case) matters in all aspects of this pipeline (parameters, column names for input files, etc.)*

## If fine-mapping must be conducted (--finemap Y)

### <ins>Required parameters</ins>

### --output_dir: 
See description above.

### --SNP_ref:
This parameter specifies the <ins>directory</ins> containing LD reference files from the 1000 Genomes project (bed, bim, fam files). These files could be downloaded from https://mrcieu.github.io/gwasglue/ (see Reference datasets subheading).

### --Population:
This is a three letter identifier for the LD reference population (e.g., EUR, AFR, EAS, AMR, SAS).

### --sum_stats:
The <ins>path</ins> to a GWAS summary statistics file. This file must be tab separated and contain at least these 4 following columns: SNP (containing SNP rsIDs), CHR (chromosome of the SNP), POS (position of SNP on chromosome), and P (p-value of association of SNP to disease). The following columns are not necessary but will be used by RegSCOUT if present: A1 (reference allele of SNP), A2 (alternative allele of SNP), MAF (minor allele frequency of SNP), and N (sample size). 

### --lead_snps: 
The <ins>path</ins> to a file containing the lead SNPs found in a GWAS study. This file must be tab separated if containing more than one column; however, only one column in this file is necessary for the pipeline to run: SNP (containing lead SNP rsIDs).

### --plink2_bin: 
This parameter is only required in the case that at least one of the A1, A2, or MAF columns is not present in the summary statistics provided to RegSCOUT. If this is the case, please provide to this parameter the <ins>path</ins> to the Plink2 executable downloaded from: https://www.cog-genomics.org/plink/2.0/. 

### --sample_num: 
This parameter can take an integer for the sample size of the GWAS (--sample_num 200000). However, if the N column (sample size) is present in the GWAS summary statistics, this column can be used instead of specifying an integer by setting this parameter as present (i.e., --sample_num present). 

### --fgwas_src:
The <ins>path</ins> to the fgwas executable after installation (e.g., --fgwas_src /home/fgwas/src/fgwas). 

### --CI_thr:
To SNPs have had their posterior probabilities of association (PPAs) assigned by fgwas, RegSCOUT finds the smallest set of SNPs in each locus whose cumulative PPAs add up to a certain threshold. This threshold is specified by this parameter by providing an integer. For example, if the smallest set of SNPs in each locus whose cumulative PPAs add up to 95% is desired then CI_thr can be set to 95 (i.e., --CI_thr 95).

### <ins>Output files</ins>

### Plink2.afreq
A file output by Plink2 which provides SNP information including reference and alternative alleles and minor allele frequency. Will not be output if A1, A2, and MAF files are present in GWAS summary statistics. 

### new_lead_snps.txt
If the A1, A2, or MAF column in missing in the GWAS summary statistics, this file will contain only those lead SNPs that were found in Plink2, which subsequent analysis will be performed on. If the three columns were present this new_lead_snps.txt file should not be different from the lead SNP file provided initially to RegSCOUT.

### loci_info.txt
File containing information on each locus region created by RegSCOUT. The columns of this file are SEGNUM (the locus number), lead_snp (the lead SNP of the locus), locus_chr (the chromosome the locus is found on), locus_start (start position of the locus), and locus_end (end position of the locus).

### final_gwas_data.txt
A tab-separated file organized by increasing locus number. This file is provided to fgwas for fine-mapping. There are 9 columns: SNPID (the rsID of a SNP), CHR (the chromosome the SNP is found on), POS (the position of the SNP on the chromosome), Z (the associated Z-score of the SNP), F (the SNP's minor allele frequency), N (sample size), SEGNUMBER (locus number), A1 (reference allele of SNP), A2 (alternative allele of SNP). 

### files starting with CI (e.g., CI.bfs.gz)
These are output files produced by fgwas as it assigns PPAs to SNPs. CI.bfs.gz contains SNPs and their associated PPA values. 

### gwas_CI.txt
A tab-separated file which contains the results of fine-mapping and filtering for CI_thr. SNPs on sex chromosomes and with PPA values <= 0.01 have also been filtered out. These are the credible interval SNPs.

## If mode is set to ATAC_obj (--mode ATAC_obj)

### <ins>Required parameters</ins>

### --output_dir: 
See description above.

### --jaspar_mtx:
In conducting TF analysis this pipeline uses the JASPAR TF binding profile database. If the user has downloaded position frequency matrices (PFMs) from JASPAR placed into a single file. This <ins>path</ins> to this file can be provided using this parameter. Alternatively, if the user would prefer not to upload such a file, this parameter can be set to none (i.e., --jaspar_mtx none). RegSCOUT in this case will obtain PFMs from JASPAR using the [JASPAR2024 library](https://bioconductor.org/packages/release/data/annotation/html/JASPAR2024.html). 

### --genome_built:
This parameter can be set to either hg19 or hg38 depending on the human genome build used in the GWAS data provided to RegSCOUT.

### --ci_gwas_dir:
If the finemap parameter was set to Y, this parameter should not be used as fine-mapping results will automatically be passed to RegSCOUT for downstream steps. If fine-mapping was not conducted using RegSCOUT, this parameter should be set to the <ins>path</ins> to the fine-mapping results. 

### --seurat_obj: 
The <ins>path</ins> to the Seurat object for the scATAC-seq data to be used in the integrative analysis. The provided Seurat object should have an Assay named "peaks" which holds a chromatin assay object created from the counts matrix of the scATAC-seq experiment. In addition, cell type labels should be provided as Idents in the Seurat object. The seurat object should be saved as an RData file. 

### --peak_th:
This parameter is used to define a threshold, above which a open chromatin region is considered to be 'open' in a cell type. For example, if --peak_th is set to 0.1 (i.e., --peak_th 0.1), then greater than 10% of cells in a cell type must have nonzero scATAC-seq read counts at an open chromatin region, for an open chromatin region to be considered 'open' in that cell type.

### --prom_th_up:
This parameter is used to define the number of bases upstream a gene's transcription start site (TSS) that will be considered part of a gene's promoter. For example, if the user wants this value to be 2000 bases upstream the TSS they would set this parameter as 2000 (i.e., --prom_th_up 2000). 

### --prom_th_down:
This parameter is used to define the number of bases downstream a gene's TSS that will be considered part of the genes promoter. See example for --prom_th_up.

### --gencode_dir:
The <ins>path</ins> to the [GENCODE gene annotation](https://www.gencodegenes.org). This gene annotation should be a gff3 file.

### <ins>Output files</ins>

### atSNP_10runs_results.RDS
An RDS file containing the results of 10 runs of the ComputePValues function of the atSNP R package which identifies the impact of SNPs on TF binding affinities. 

### Ci_effect_SNPs.txt
A tab-separated file containing all identified effect-SNPs and their associated TFs. Effect-SNPs are SNPs which significantly affected the binding affinity of at least one TF. This file has 9 columns: SNP (the rsID of a SNP), CHR (the chromosome the SNP is found on), Pos (the position of the SNP on the chromosome), Locus (locus number), TF (TF whose binding affinity was significantly impacted by SNP), Raw_P_value (p-value before FDR correction), FDR_corr_P_value (p-value after FDR correction), log_like_ratio (the log likelihood ratio illustrating the impact of the alternative allele of a SNP on TF binding compared to the reference allele), and ppa (PPA of SNP).

### cell_peak.xlsx 
Table that provides genomic locations of peaks considered open in certain cell types (i.e., those that met the CI_thr). This table will have four columns: chr (the chromosome the open chromatin region is found on), start (start position of open chromatin region), end (end position of open chromatin region), and cell sub-types (all cell types in which a peak was considered open).

### [cell type name].filtered_coaccessible_sites4s.txt
These files are the results of cicero analysis. They will have three columns: Peak1, Peak2, and coaccess (the co-accessibility value calculated between Peak1 and Peak2). Only peak-peak interactions with co-accessibility > 0.05 will be found in these files. One file will be output for each cell type in the scATAC-seq data. 

### risk_regions_ppa.xlsx
Table that provides information on risk-mediating peaks (RMPs) found by RegSCOUT. RMPs are those open chromatin regions that contain at least one effect-SNP. It has three columns: region (the RMP), cell (the cell types in which the region was found to be an RMP), and sumPPA (the sum of the PPAs of all effect-SNPs that fall in that RMP).

### peak_cluster_matrix.txt
A binary matrix that contains cell types as columns and RMPs as rows. A '1' indicates that that a RMP was found in a specified cell type whereas a '0' means it was not. 

### cell_peak.png
A visual representation of risk_regions_ppa.xlsx. This heatmap shows RMPs on the y-axis and cell types on the x-axis. A heatmap cell shaded in red indicates that a certain RMP was found in a certain cell type whereas if a cell is shaded in white, that RMP was not found in that cell type. The bar on the right of this heatmap with shades of blue shows the sumPPA value obtained for each RMP.

### risk_regions_ratio.xlsx
Table providing information on RMPs and their associated effect-SNPs. This table has 4 columns: TFSNP (a TF-SNP pair identified by atSNP), region (an RMP), cell (the cell type in which this effect SNP was found to fall in this RMP), and log_lik_ratio (the log likelihood ratio describing an effect-SNPs effect on the TF's binding affinity).

### TF_cluster_matrix.txt
A binary matrix that contains cell types as columns and TF-SNP pairs as rows. If a TF-SNP pair was found relevant to a certain cell type it would obtain a value of '1', otherwise it would obtain a value of '0'.

### cell_tf.png
A visual representation of risk_regions_ratio specific to TF-SNP pairs. This heatmap shows TF-SNP pairs on the y-axis and cell types on the x-axis. A heatmap cell shaded in red indicates that a certain TF-SNP pair was found relevant to a certain cell type whereas if a cell is shaded in white, that TF-SNP pair was not found relevant to that cell type. The bar on the right of this heatmap with shades of blue shows the log likelihood ratio value obtained for each TF-SNP pair.

### Aff_peak_interact_gene.csv
Table that shows the links between RMPs and their target genes. This table has 6 columns: Peak1 (An RMP), Peak2 (the co-accessible peak associated with the RMP), coaccess (the co-accessibility value between those two peaks), Promoter (the promoter region of the affected gene defined by --prom_th_up and --prom_th_down parameters), Gene (the id of affected gene), and Cell_Type (the cell type in which this putative RMP-gene interaction was detected). Please note that the rows with identical Peak1 and Peak2 represent the RMPs which directly overlap the promoters of their target genes.

### cell_gene.csv
A binary matrix that shows genes as columns and cell types as rows. A '1' indicates that a gene was prioritized in a cell type according to cicero (co-accessibility value > 0.05 between an RMP and the gene's promoter), whereas a value of '0' indicates the gene was not prioritized in that cell type. 

### cell_gene.png
A visual representation of cell_gene.csv. This heatmap shows cell types on the y-axis and genes on the x-axis. A heatmap cell shaded in red indicates that a certain gene was prioritized to a certain cell type whereas if a cell is shaded in white, that gene was not found relevant to that cell type according to cicero.

### Gene_promoter_matrix.txt
A binary matrix that shows cell types as columns and RMP-gene pairs as rows. This matrix is specific to those genes whose promoters were overlapped by an RMP. A '1' indicates that that RMP-gene pair was identified in that cell type, whereas a '0' indicates that RMP-gene pair was not found relevant to that cell type. 

### cell_gene_promoter.png
A visual representation of Gene_promoter_matrix.txt. This heatmap is specific to those genes whose promoters were overlapped by an RMP. This heatmap shows RMP-gene pairs on the y-axis and cell types on the x-axis. A heatmap cell shaded in red indicates that a certain RMP-gene pair was found relevant to a certain cell type whereas if a cell is shaded in white, that pair was not found relevant to that cell type.

### Gene_enhancer_matrix.txt
A binary matrix that shows cell types as columns and RMP-gene pairs as rows. This matrix is specific to the case where cicero identified a relationship between an RMP and a gene's promoter where the RMP did not overlap the gene's promoter. A '1' indicates that that RMP-gene pair was identified in that cell type, whereas a '0' indicates that RMP-gene pair was not found relevant to that cell type. 

### cell_gene_enhancer.png
A visual representation of Gene_enhancer_matrix.txt. This heatmap is specific to those genes whose promoters were found to be co-accessible with an RMP but the RMP did not directly overlap the promoter. This heatmap shows RMP-gene pairs on the y-axis and cell types on the x-axis. A heatmap cell shaded in red indicates that a certain RMP-gene pair was found relevant to a certain cell type whereas if a cell is shaded in white, that pair was not found relevant to that cell type.

## If mode is set to peak_table (--mode peak_table)

### <ins>Required parameters</ins>
Parameters required for --mode peak_table are nearly the same as those required for --mode ATAC_obj; the difference is that the --seurat_obj and --peak_th parameters are no longer necessary. The peak and cell type table and peak-peak interaction tables should be placed in the directory specified by --output_dir. 

### <ins>The peak and cell type table, and peak-peak interaction tables</ins>
This peak and cell type table should be named cell_peak.xlsx and should have three columns named "chr", "start", "end" (chromosome, start and end of the open chromatin regions), and a column named “cell sub-types” which represents the cell types in which a peak is considered open. Multiple cells should be comma-separated in this column. There should be one peak-peak interaction table for each cell type to be studied. The names of their associated files should be [cell type name].filtered_coaccessible_sites4s.txt. These tables should be tab-separated and should have three columns named "Peak1", "Peak2", and "coaccess" (coaccessibility value between peak1 and peak2 calculated by cicero). The Peak1 and Peak2 columns should contain underscore-separated locations of open chromatin regions (e.g., chr1_1000_1600). Examples of a peak and cell type table and of a peak-peak interaction table are shown below. 

Example peak and cell type table:
| chr | start | end | cell sub-types | 
| ---- | ----- | --- | -------------- | 
| chr1 | 123 | 1000 | plasma_cells,memory_b_cells | 
| chr11 | 2000 | 2600 | naive_CD8_T_cells,naive_CD4_T_cells | 
| chr18 | 1 | 678 | classical_monocytes,intermediate_monocytes,nonclassical_monocytes | 

Example peak-peak interaction table:
| Peak1 | Peak2 | coaccess | 
| ---- | ---- | ----- | 
| chr1_1000_1600 | chr1_2000_2600 | 0.13 |
| chr3_3000_4000 | chr3_5000_5500 | 0.27 |

### <ins>Output files</ins>
The output for this mode would be the same as those listed in --mode ATAC_obj; the difference is that cell_peak.xlsx and [cell type name].filtered_coaccessible_sites4s.txt are not output as they are required as inputs when --mode is set to peak_table. 
