---
output:
  pdf_document: default
  html_document: default
  word_document: default
---
![logo](https://github.com/shooshtarilab/RegSCOUT/blob/main/images/RegSCOUT_logo.png)

# RegSCOUT (<ins>Reg</ins>ulatory <ins>S</ins>ingle <ins>C</ins>ell <ins>O</ins>mics for <ins>U</ins>nravelling <ins>T</ins>rait Loci - Updates in Progress)
This is the repository for the RegSCOUT pipeline. This pipeline can be used as a stand-alone command-line application which integrates disease-associated SNPs, identified through genome-wide association studies (GWAS), with scATAC-seq, Hi-C, and eQTL data. Through this integrative analysis, RegSCOUT prioritizes putative gene regulatory networks relevant to complex disease pathogenesis. Example bash scripts for use of RegSCOUT can be found in the bash_scripts folder on this repository. 

This pipeline can handle both a scATAC-seq data as a Seurat object or processed results of analyzing scATAC-seq data as a peak and cell type table along with peak-peak interaction tables (generated by [cicero](https://bioconductor.org/packages/release/bioc/html/cicero.html)). See the --mode peak_table section for more information about these two tables. 

## Required R version and libraries:
RegSCOUT was created using R version 4.4.0 and uses these libraries: Seurat, Signac, GenomicRanges, genetics.binaRies, R.utils, atSNP, TFBSTools, stringr, circlize, ComplexHeatmap, xlsx, R.utils, ape, monocle3, cicero, Repitools, reshape2, dplyr, BSgenome.Hsapiens.UCSC.hg19, BSgenome.Hsapiens.UCSC.hg38, ggplot2, readxl, tidyr, tibble. Monocle3 must be installed prior to installing cicero. Cicero should be installed following instructions found in this link: https://cole-trapnell-lab.github.io/cicero-release/docs_m3/. 

## Note about genome builds:
When using RegSCOUT please ensure that all files provided to this pipeline are in the same genome build. E.g., PLINK bfiles (.bed, .bim, .fam), GWAS summary statistics, GWAS lead SNPs, GWAS locus region information, fine-mapped GWAS data, scATAC-seq datasets, Hi-C datasets, eQTL datasets, histone mark results datasets, and GENCODE gene information. 

## Pipeline Overview
![Pipeline Overview](https://github.com/shooshtarilab/RegSCOUT/blob/main/images/Workflow_schematic.png)

Above is a general overview of the RegSCOUT workflow. Each step will be explained below with a description of inputs and outputs. Although these steps are numbered 1-10, many are optional and dependent on what datasets are available to the user:
- If --mode *peak_table* is requested, step 3 will not be run. 
- If the --finemap parameter is not specified, steps 1-2 will not run. 
- If the --hic_analysis parameter is not specified, step 6 will not run.
- If the --eqtl_analysis parameter is not specified, step 7 will not run.
- If the --histone_mark_analysis parameter is not specified, step 8 will not run.
- If the --tf_expr_analysis parameter is not specified, step 9 will not run.

A table of all parameters included in RegSCOUT including which steps the parameter is relevant to, potential values, default value, and general purpose of the parameter can also be found below.

## Table of All Parameters
| Parameter | Purpose and Potential Values | Default Value | Relevant Steps |
| --------------------- | ----------------------------------------------------- | ---------------- | ---------------------- |
| --snp_ref_dir | Specifies the directory containing the bfiles used for PLINK preprocessing of GWAS data and LD analysis. | No default, must be set by user. | 1 |
| --population | Specifies the population/ancestry that the bfiles were created from (e.g., EUR, EAS). | No default, must be set by user. | 1 |
| --sum_stats_dir | Specifies the path to the GWAS summary statistics. | No default, must be set by user | 1 |
| --lead_snps_dir | Specifies the path to GWAS lead SNP information. | No default, must be set by user | 1 |
| --plink2_dir | Specifies the path to the Plink2 executable. | No default, must be set by user | 1 |
| --sample_num | Sample size (total number of cases and controls) of the GWAS. If a sample size column (N) with sample size for each SNP is provided in the GWAS summary statistics, this parameter should be set to *present*. Alternatively, an integer value can be provided in this parameter to specify the overall sample size of the GWAS, this should only be provided if a sample size for each SNP is not readily available. | *present* | 1 |
| --locus_region | The # of bps to add on both sides of lead SNP positions to define locus regions. | 1,000,000 | 1 |
| --ld_th | SNPs are not included in a locus if they obtain an abs(LD value) (absolute value), with the lead SNP, less than or equal to this threshold. | 0.25 | 1 |
| --fgwas_dir | Specifies the path to the fgwas executable. | No default, must be set by user | 2 |
| --ci_th | For each locus, RegSCOUT filters for the smallest group of SNPs whose cumulative posterior probabilities of association (PPAs) add up to this threshold. | 0.95 | 2 |
| --ci_ppa_th | Each CI SNP must have a PPA greater than this threshold. | 0.01 | 2 |
| --finemap | Set this parameter to *Y* if fine-mapping should be conducted by RegSCOUT using [fgwas](https://github.com/joepickrell/fgwas). If fine-mapping is not desired (i.e., the user already has fine-mapping results), this parameter should not be used. | Not conducted | 1, 2 |
| --seurat_obj_dir | Specifies the path to the scATAC-seq Seurat object. | No default, must be set by user | 3 |
| --cell_count_th | Cell types with less than this number of cells in the scATAC-seq dataset will be removed. | 3 | 3 |
| --coaccess_th | The co-accessibility value between two peaks must be greater than this threshold to be output in cicero results. | 0.05 | 3 |
| --cic_genomic_window | Size of the regions on either side of an open chromatin region that cicero will look for peak to peak links in. | 2,000,000 | 3 |
| --jaspar_mtx_dir | Specifies the path to a user provided file with JASPAR TF PFMs. If set to *none*, the JASPAR2024 R library will be used to obtain TF PFMs. | *none* | 4 |
| --ci_gwas_dir | Only use this parameter if fine-mapping <ins>was not</ins> conducted by RegSCOUT. Specifies the path to fine-mapping results. | If fine-mapping was not conducted by RegSCOUT: no default, must be set by user | 4 |
| --ncores | Number of cores to use when computing TF-SNP pair p-values using atSNP. Generally a greater number of cores means faster computing time. | 2 | 4 |
| --genome_build | The reference genome that RegSCOUT should conduct analyses in, either hg19 or hg38. | No default, must be set by user | 3, 4 |
| --peak_th | The threshold above which a peak is considered to be open in a cell type. E.g., a peak can be considered open if it is accessible in 1/10 or 10% of cells of a cell type. | 0.1 | 3, 4 |
| --prom_th_up | Number of bps to add upstream a TSS to define the gene promoter. | 2,000 | 5 |
| --prom_th_down | Number of bps to add downstream a TSS to define the gene promoter. | 2,000 | 5 |
| --gencode_dir | Specifies the path to GENCODE gene information. | No default, must be set by user | 5 |
| --mode | The user must indicate whether they have a scATAC-seq Seurat object or if they have a table of open chromatin regions as well as cicero results already generated. If the user is providing a seurat object, this parameter should be set to *ATAC_obj*. If the user is providing a peak table and cicero results, this parameter should be set to *peak_table*. | No default, must be set by user | 3, 4, 5 |
| --hic_analysis | Set this parameter to *Y* if gene regulatory analysis using Hi-C analysis is desired. If Hi-C analysis is not desired, do not use this parameter | Analysis not conducted | 6 |
| --hic_instruct_dir | Specifies the path to user generated Hi-C analysis instructions. | No default, must be set by user | 6 |
| --eqtl_analysis | Set this parameter to *Y* if gene regulatory analysis using eQTL analysis is desired. | Analysis not conducted | 7 |
| --eqtl_instruct_dir | Specifies the path to user generated eQTL analysis instructions. | No default, must be set by user | 7 |
| --histone_mark_analysis | Set this parameter to *Y* if analysis of risk-mediating peaks using user-provided histone mark data is desired | Analysis not conducted | 8 |
| --hist_mark_instruct_dir | Specifies the path to user generated histone mark analysis instructions. | No default, must be set by user | 8 |
| --tf_expr_analysis | Only use this parameter if TF expression analysis using scATAC-seq, scRNA-seq, or both is desired. Set to *atac* if TF expression analysis is desired using scATAC-seq data that was provided as input earlier for --mode *peak_table* or *ATAC_obj*. Set to *rna* if analysis is desired using user-provided scRNA-seq data. Set to *both* if both scATAC-seq and scRNA-seq analyses are desired. | No conducted | 9 |
| --scrna_instruct_dir | Specifies the path to user generated scRNA-seq analysis instructions spreadsheet. | No default, must be set by user | 9 |
| --tf_rna_quantile_th | The quantile for percent expression of genes in a cell type, above which a TF will be considered expressed in that cell type. | 0.25 | 9 |
| --tf_score_th | Genes will only be prioritized if one of their associated TFs attains a score higher than this threshold. | -1 | 10 |
| --gene_sum_ppa_th | Genes will only be prioritized if their sum ppa value is greater than this threshold. | 0.05 | 10 |
| --gene_score_th | Genes will only be prioritized if their gene score is greater than this threshold. | 1 | 10 |
| --output_dir | The directory in which all RegSCOUT output files will be stored. | No default, must be set by user | 1 through 10 |

*All directories specified in parameters should end with "/".

**Please note that case (lower/upper case) matters in all aspects of this pipeline (parameters, column names for input files, etc.).

***When providing numerical values to parameters, do not incorporate commas or spaces for thousands, e.g., use 100000 not 100,000.

****All chromosome numbers provided in files to RegSCOUT must be of the following naming convention 'chr' + 'number' (no space between). For example: chr8.




## Main Parameters:
### --mode: 
This parameter should be used to choose the mode of operation of the pipeline based on input type. If a scATAC-seq Seurat object is provided, this option should be set to ATAC_obj (i.e., --mode ATAC_obj). If a peak and cell type table and peak-peak interaction table are used this mode should be set to peak_table (i.e., --mode peak_table).

### --output_dir: 
This parameter should be used for setting the working directory (the directory address should end with "/"). **This parameter should always be set when using RegSCOUT regardless of the value of --finemap or --mode**. All the output and intermediate files of the pipeline are generated in this directory. If running the pipeline with --mode peak_table, the peak-by-cell and peak-peak interaction tables of the pipeline should be provided in this directory.

## Fine-mapping Requested by User (--finemap Y; Steps 1 and 2)
See required parameters in table of all parameters.

### <ins>Required Files</ins>

### GWAS Summary Statistics
The path to the summary statistics file should be specified using --sum_stats_dir. This file must be tab separated and contain at least these 4 following columns (note that RegSCOUT is case sensitive): SNP (containing SNP rsIDs), CHR (chromosome of the SNP), POS (position of SNP on chromosome), and P (p-value of association of SNP to disease). The following columns are not necessary but will be used by RegSCOUT if present: A1 (reference allele of SNP), A2 (alternative allele of SNP), MAF (minor allele frequency of SNP), and N (sample size). 

If a sample size column is present in the summary statistics, the --sample_num parameter should be set to *present*; however, if it is not and the user is not able to find sample size for each SNP, the overall sample size of the GWAS study should be provided using the --sample_num parameter (e.g., --sample_num 190000). 

If one of A1, A2, or MAF are missing from the GWAS summary statistics file, each of these columns will be filled in using Plink2 and its required bfiles. More information regarding this can be found in the Plink bfiles section. 

### GWAS Lead SNPs
The path to this file should be specified using --lead_snp_dir. This file must be tab separated (.txt or .tsv) if containing more than one column; however, only one column in this file is necessary for the pipeline to run: SNP (containing lead SNP rsIDs). Any additional columns provided will be included in the new_lead_snps.txt output file described in the "Output Files" section. 

### Plink bfiles
The path to these files are specified by the combination of the --snp_ref_dir and --population parameters. The file suffixes are not required in the file path, thus an example file path used by RegSCOUT is '/user/bin/EUR'. These are the .bim, .bed, .fam files required to run Plink2 to fill in reference/alternative allele as well as minor allele frequency information. These files will also be used to calculate LD values and group SNPs into loci. These files in hg19 can be downloaded from here: https://mrcieu.github.io/gwasglue/. Users can also provide bfiles that they have generated, in hg38, or for the specific group they want to study.

### <ins>Output Files</ins>

### Plink2.afreq
A file output by Plink2 which provides SNP information including reference and alternative alleles and minor allele frequency. Will not be output if all of A1, A2, and MAF columns are present in GWAS summary statistics.

### new_lead_snps.txt
If the A1, A2, or MAF column in missing in the GWAS summary statistics, this file will contain only those lead SNPs that were found in Plink2, which subsequent analyses will be performed on. If the three columns were present this new_lead_snps.txt file should not be different from the lead SNP file provided initially to RegSCOUT.

### loci_info.txt
File containing information on each locus region created by RegSCOUT. The columns of this file are chunk (the locus number), lead_snp (the lead SNP of the locus), locus_chr (the chromosome the locus is found on), locus_start (start position of the locus), and locus_end (end position of the locus). 

### final_gwas_data.txt
A tab-separated file organized by increasing locus number. This file is provided to fgwas for fine-mapping. There are 9 columns: SNPID (the rsID of a SNP), CHR (the chromosome the SNP is found on), POS (the position of the SNP on the chromosome), Z (the associated Z-score of the SNP), F (the SNP's minor allele frequency), N (sample size), SEGNUMBER (locus number), A1 (reference allele of SNP), A2 (alternative allele of SNP). 

### gwas_CI.txt
A tab-separated file which contains the results of fine-mapping and filtering for CI_thr. SNPs on sex chromosomes and with PPA values <= 0.01 have also been filtered out. These are the credible interval SNPs.

## Fine-mapping Already Conducted (--finemap not used)

### <ins>Required Files</ins>

### Fine-mapping Results File
The path to this file must be specified by the --ci_gwas_dir parameter. If fine-mapping has already been conducted by the user, a file that contains the results of fine-mapping is required to be provided by the user. This file must contain credible interval SNPs where for each locus, the smallest number of SNPs whose cumulative PPAs add up to a certain threshold is provided (e.g., 0.99 or 0.95, this threshold is up to the user). This identification of credible interval SNPs will not be conducted by RegSCOUT if the --finemap parameter is not used. The file does not have a specific naming convention that it must follow, but must be a tab separated file (.txt or .tsv). This file must have at least 7 columns labelled: 'id' (SNP rsID), 'pos' (SNP bp position), 'chr' (SNP chromosome), 'chunk' (SNP locus number), 'PPA' (SNP associated PPA), 'a1' (SNP reference allele), and 'a2' (SNP alternative allele). Other columns that are optional but if present will be included in the final tables generated are: 'lead_snp' (the lead SNP of the SNP's locus), 'locus_chr' (locus chromosome number e.g., chr8 - chromosome numbers must follow this naming convention), 'locus_start' (start position of the locus), and 'locus_end' (end position of the locus).

## Mode set to peak_table (--mode peak_table; Steps 4 and 5)
The *peak_table* option should only be used if the user has already identified cell type-specific open chromatin regions and has already run cicero on their data to establish peak to peak links. If the user has specified this option for RegSCOUT, this means that RegSCOUT will not be identifying cell type-specific open chromatin regions based on a certain threshold (e.g., a region must be open in 10% of cells of a cell type). RegSCOUT will also not be running cicero analysis to link open chromatin regions to each other as that must have already been done by the user. Instead, the user must provide both a peak-cell type table and cell type-specific co-accessibility tables (output of cicero) to RegSCOUT **in the directory specified by the --output_dir parameter**. More information on these tables and their formats can be found below.

### <ins>Required Files</ins>

### Peak-Cell Type Table
The path to this file need not be specified as RegSCOUT will look for this file in the directory defined by the --output_dir parameter. This file must be a tab separated file (.tsv) named **cell_peak.tsv**. The table should have four columns "chr", "start", "end", and "cell_types". Each row will represent one region of chromatin that was identified to be open. The "cell_types" column in each row will be a comma separated list of all cell types in which the chromatin region on that row was identified to be open. In the 'cell_types' column, there should be no spaces after each comma. 

Example peak-cell type table:
| chr | start | end | cell_types | 
| ---- | ----- | --- | -------------- | 
| chr1 | 123 | 1000 | plasma_cells,memory_b_cells | 
| chr11 | 2000 | 2600 | naive_CD8_T_cells,naive_CD4_T_cells | 
| chr18 | 1 | 678 | classical_monocytes,intermediate_monocytes,nonclassical_monocytes | 

### Cell Type-Specific Co-Accessibility Tables
The path to these files need not be specified as RegSCOUT will look for these files in the directory defined by the --output_dir parameter. These should be the output of cicero after filtering for a certain co-accessibility threshold. RegSCOUT will not perform this co-accessibility filtering if --mode *peak_table* was selected. There will be one table for each of the cell types investigated. The names of each cell type-associated file should follow this naming convention: [cell type name].filtered_coaccessible_sites4s.txt. These tables should be tab-separated and should have three columns named "Peak1", "Peak2", and "coaccess" (coaccessibility value between peak1 and peak2 calculated by cicero). The Peak1 and Peak2 columns should contain underscore-separated locations of open chromatin regions (e.g., chr1_1000_1600). 

Example cell type-specific co-accessibility table:
| Peak1 | Peak2 | coaccess | 
| ---- | ---- | ----- | 
| chr1_1000_1600 | chr1_2000_2600 | 0.13 |
| chr3_3000_4000 | chr3_5000_5500 | 0.27 |

### GENCODE Gene Annotation File
The path to this file should be specified using the --gencode_dir parameter. The <ins>path</ins> to the GENCODE gene annotation file downloaded from the [GENCODE website](https://www.gencodegenes.org). Users should download the appropriate .gff3 file in the genome build that all other files are in, that the user is hoping to have RegSCOUT analyze. The .gff3 file should be a comprehensive gene annotation file. RegSCOUT will filter this gene annotation file for specifically the protein coding gene transcripts present and use those transcripts for subsequent analyses, i.e., defining promoter regions.

### JASPAR TF PFM File
The path to this file should be specified using the --jaspar_mtx_dir parameter. To create this file, the user must download position frequency matrices (PFMs) from [the JASPAR website](https://jaspar.elixir.no) and place them into a single file. In conducting TF analysis, this pipeline uses the JASPAR TF binding profile database. This <ins>path</ins> to this file can be provided using this parameter. Alternatively, if the user would prefer not to upload such a file, this parameter can be set to none (i.e., --jaspar_mtx none). RegSCOUT in this case will obtain PFMs from JASPAR using the [JASPAR2024 library](https://bioconductor.org/packages/release/data/annotation/html/JASPAR2024.html). 

### <ins>Output Files</ins>

### atSNP_10runs_results.RDS
An RDS file containing the results of 10 runs of the ComputePValues function of the atSNP R package which identifies the impact of SNPs on TF binding affinities. 

### Ci_effect_SNPs.txt
A tab-separated file containing all identified SNPs significantly impacting the binding affinity of at least one TF (<ins>termed Effect-SNPs</ins>), and their associated TFs. This file has 9 columns: 'SNP' (the rsID of a SNP), 'CHR' (the chromosome the SNP is found on), 'Pos' (the position of the SNP on the chromosome), 'Locus' (locus number), 'TF' (TF whose binding affinity was significantly impacted by SNP), 'Raw_P_value' (p-value before FDR correction), 'FDR_corr_P_value' (p-value after FDR correction), 'log_like_ratio' (the log likelihood ratio quantifying the impact of the alternative allele of a SNP on TF binding affinity compared to the reference allele), and 'ppa' (PPA of SNP).

### peak_cluster_matrix.txt
A binary matrix that contains cell types as columns and open chromatin regions that co-localize with at least one effect-SNP as rows (<ins>these regions are termed risk-mediating peaks</ins>). A '1' indicates that that an RMP was found in a specified cell type whereas a '0' means it was not. This text file is tab separated.

### risk_regions_ppa.txt
Table that provides more information on risk-mediating peaks found by RegSCOUT. It has three columns: 'region' (the RMP), 'cell' (the cell types in which the open chromatin region was found to be an RMP), and sumPPA (the sum of the PPAs of all effect-SNPs that fall in that RMP).

### cell_peak.svg
A visual representation of risk_regions_ppa.txt. This heatmap shows risk-mediating peaks on the y-axis and cell types on the x-axis. A heatmap square shaded in red indicates that a certain risk-mediating peak was found in a certain cell type whereas if a square is shaded in white, that RMP was not found in that cell type. The bar on the right of this heatmap with shades of purple shows the summed PPA (sumPPA) value obtained for each risk-mediating peak.

### TF_cluster_matrix.txt
A binary matrix that contains cell types as columns and TF-SNP pairs as rows. If a TF-SNP pair was found relevant to a certain cell type it would obtain a value of '1', otherwise it would obtain a value of '0'.

### risk_regions_ratio.txt
Table providing information on risk-mediating peaks and their associated effect-SNPs. This table has 4 columns: 'region' (a risk-mediating peak), 'cell' (the cell type in which this effect SNP was found to fall in this risk-mediating peak), 'TFSNP' (a TF-SNP pair identified by atSNP), and 'log_like_ratio (the log likelihood ratio describing an effect-SNPs effect on a TF's binding affinity).

### cell_snp_tf.svg
A visual representation of risk_regions_ratio.txt specific to TF-SNP pairs. This heatmap shows TF-SNP pairs on the y-axis and cell types on the x-axis. A heatmap square shaded in blue indicates that a certain TF-SNP pair was found relevant to a certain cell type whereas if a square is shaded in white, that TF-SNP pair was not found relevant to that cell type. The bar on the right of this heatmap with shades of orange shows the log likelihood ratio value obtained for each TF-SNP pair.

### direct_rmp_gene_overlaps.txt
This is a table of all risk-mediating peak and gene promoter (as defined by GENCODE) overlaps identified by RegSCOUT, output as a tab separated text file. There are 4 columns: 'RMP' (the risk-mediating peak), 'Promoter' (the promoter region defined using GENCODE TSSs, and --prom_th_up and --prom_th_down parameters defined by the user), 'Gene' (the gene associated with that promoter), and 'Cell_Type' (the cell type in which that risk-mediating peak was identified).

### cic_peak_interact_gene.txt
Table that shows links between RMPs and their target genes using peak-peak links identified by cicero, output as a tab separated text file. This table has 6 columns: 'Peak1' (A risk-mediating peak), 'Peak2' (The peak that co-localized with a gene's promoter), 'coaccess' (the coaccessibility value calculated by cicero between Peak1 and Peak2), 'Promoter' (the promoter region defined using GENCODE TSSs, and --prom_th_up and --prom_th_down parameters defined by the user), 'Gene' (the gene associated with that promoter), and 'Cell_Type' (the cell type in which this RMP-gene link was identified).

### cell_gene_matrix.txt
A binary matrix that shows genes as columns and cell types as rows. A '1' indicates that a gene was prioritized in a cell type according to risk-mediating peak and gene promoter overlap with cicero results, or by direct co-localization of a risk-mediating peak with a gene's promoter. A '0' indicates that a gene was not prioritized in a cell type according to those criteria. 

### Gene_promoter_matrix.txt
A binary matrix that shows cell types as columns and RMP-gene pairs as rows. This matrix is specific to those genes whose promoters were overlapped by an RMP. A '1' indicates that that RMP-gene pair was identified in that cell type, whereas a '0' indicates that RMP-gene pair was not found relevant to that cell type. 

### Gene_enhancer_matrix.txt
A binary matrix that shows cell types as columns and RMP-gene pairs as rows. This matrix is specific to the case where cicero identified a relationship between an RMP and a gene's promoter where the RMP did not overlap the gene's promoter. A '1' indicates that that RMP-gene pair was identified in that cell type, whereas a '0' indicates that RMP-gene pair was not found relevant to that cell type. 

## Mode set to ATAC_obj (--mode ATAC_obj; Steps 3, 4, and 5)
The *ATAC_obj* option should be used when the user has a scATAC-seq dataset that they would like RegSCOUT to use to identify risk-mediating open chromatin regions and run cicero analysis. More information on how this scATAC-seq dataset should be formatted can be found below. In this case, RegSCOUT will identify open chromatin regions in the scATAC-seq dataset based on a threshold defined using the --peak_th parameter. RegSCOUT will also run cicero analysis according to the user's specifications using the --cic_genomic_window and --coaccess_th parameters. 

### <ins>Required Files</ins>
The required files when using --mode ATAC_obj are different from those required for --mode peak_table with the exception of the GENCODE gene annotation file, which is still required and the path to this file should still be set with the --gencode_dir parameter. 

### scATAC-seq Seurat Object
The path to the scATAC-seq Seurat object should be provided using the --seurat_obj_dir parameter. The provided Seurat object should have an Assay named "peaks" which holds a chromatin assay object created from the counts matrix of the scATAC-seq experiment. In addition, cell type labels should be provided as cell identities in the Seurat object (and accessible using the Idents() function of Seurat). This article provides more information on how to establish cell identities/class labels (https://satijalab.org/seurat/articles/essential_commands.html). The seurat object can be provided to RegSCOUT as either an RDS or RData file, both will work with the pipeline.

### JASPAR TF PFMs and GENCODE Gene Annotation

### <ins>Output files</ins>

### cell_peak.xlsx 
Table that provides genomic locations of peaks considered open in certain cell types (i.e., those that met the CI_thr). This table will have four columns: chr (the chromosome the open chromatin region is found on), start (start position of open chromatin region), end (end position of open chromatin region), and cell sub-types (all cell types in which a peak was considered open).

### [cell type name].filtered_coaccessible_sites4s.txt
These files are the results of cicero analysis. They will have three columns: Peak1, Peak2, and coaccess (the co-accessibility value calculated between Peak1 and Peak2). Only peak-peak interactions with co-accessibility > 0.05 will be found in these files. One file will be output for each cell type in the scATAC-seq data. 

### cell_gene.png
A visual representation of cell_gene.csv. This heatmap shows cell types on the y-axis and genes on the x-axis. A heatmap cell shaded in red indicates that a certain gene was prioritized to a certain cell type whereas if a cell is shaded in white, that gene was not found relevant to that cell type according to cicero.

## If mode is set to peak_table (--mode peak_table)

### <ins>Required parameters</ins>
Parameters required for --mode peak_table are nearly the same as those required for --mode ATAC_obj; the difference is that the --seurat_obj and --peak_th parameters are no longer necessary. The peak and cell type table and peak-peak interaction tables should be placed in the directory specified by --output_dir. 

### <ins>The peak and cell type table, and peak-peak interaction tables</ins>
This peak and cell type table should be named cell_peak.xlsx and should have three columns named "chr", "start", "end" (chromosome, start and end of the open chromatin regions), and a column named “cell sub-types” which represents the cell types in which a peak is considered open. Multiple cells should be comma-separated in this column. There should be one peak-peak interaction table for each cell type to be studied. The names of their associated files should be [cell type name].filtered_coaccessible_sites4s.txt. These tables should be tab-separated and should have three columns named "Peak1", "Peak2", and "coaccess" (coaccessibility value between peak1 and peak2 calculated by cicero). The Peak1 and Peak2 columns should contain underscore-separated locations of open chromatin regions (e.g., chr1_1000_1600). Examples of a peak and cell type table and of a peak-peak interaction table are shown below. 

### <ins>Output files</ins>
The output for this mode would be the same as those listed in --mode ATAC_obj; the difference is that cell_peak.xlsx and [cell type name].filtered_coaccessible_sites4s.txt are not output as they are required as inputs when --mode is set to peak_table. 
