![logo](https://github.com/shooshtarilab/RegSCOUT/blob/main/Logo/RegSCOUT_logo.png)

# RegSCOUT (<ins>Reg</ins>ulatory <ins>S</ins>ingle <ins>C</ins>ell <ins>O</ins>mics for <ins>U</ins>nravelling <ins>T</ins>rait Loci)(README has not been fully updated)
This is the repository for the RegSCOUT pipeline. This pipeline can be used as a stand-alone command line application which integrates disease-associated SNPs, identified through genome-wide association studies (GWAS), with scATAC-seq, Hi-C, and eQTL data. This pipeline can handle both a scATAC-seq data as a Seurat object or processed results of analyzing scATAC-seq data as a peak and cell type table (see below example of such a table) along with peak-peak interaction tables (generated by [cicero](https://bioconductor.org/packages/release/bioc/html/cicero.html)). 

Example peak and cell type table, it must have these 4 columns:
| chr | start | end | cell sub-types | 
| ---- | ----- | --- | -------------- | 
| chr1 | 123 | 1000 | plasma_cells,memory_b_cells | 
| chr11 | 2000 | 2600 | naive_CD8_T_cells,naive_CD4_T_cells | 
| chr18 | 1 | 678 | classical_monocytes,intermediate_monocytes,nonclassical_monocytes | 

Descriptions of required parameters and of RegSCOUT output can be found below. 

## Main Parameters:
### --mode: 
This parameter should be used to choose the mode of operation of the pipeline based on input type. If a scATAC-seq Seurat object is provided, this option should be set to ATAC_obj (i.e., --mode ATAC_obj). If peak-by-cell and peak-peak interaction tables are used this mode should be set to peak_table (i.e., --mode peak_table).

In this case, the path to the Seurat object should be set with the --seurat_obj parameter. In this case, these two tables should be provided in the working directory (set by --output_dir parameter). 

This table should be exactly named as cell_peak.xlsx and should have three columns named “chr”, “start”, “end” (chromosome, start and end of the peaks), and “cell sub-types” which represents the cell types in which a peak is active. Multiple cells should be comma-separated in this column. **This table should not have any heading and column names should be in the first line.** In addition, cell type-specific peak-peak interaction tables should be provided in the working directory. The file names should be [cell type names].filtered_coaccessible_sites4s.txt. These tables should have three columns exactly named peak1, peak2, and coaccess (coaccessibility value between Peak1 and peak2). 

### --finemap:
This parameter should be used only when the user wants fine-mapping to be conducted within the RegSCOUT workflow using [fgwas](https://github.com/joepickrell/fgwas). If this is the case --finemap should be set to Y (i.e., --finemap Y). If this is not the case, this parameter should not be used. 

### --output_dir: 
This parameter should be used for setting the working directory (the directory address should end with "/"). **This parameter should always be set when using RegSCOUT regardless of the value of --finemap or --mode**. All the output and intermediate files of the pipeline are generated in this directory. If running the pipeline with --mode peak_table, the peak-by-cell and peak-peak interaction tables of the pipeline should be provided in this directory.

*All directories specified in parameters should end with "/"*

*Please note that case (lower/upper case) matters in all aspects of this pipeline (parameters, column names for input files, etc.)*

## If fine-mapping must be conducted (--finemap Y)

### *Required parameters*

### --output_dir: 
See description above.

### --SNP_ref:
This parameter specifies the <ins>directory</ins> containing LD reference files from the 1000 Genomes project (.bed, .bim, .fam files). These files could be downloaded from https://mrcieu.github.io/gwasglue/ (see Reference datasets subheading).

### --Population:
This is a three letter identifier for the LD reference population (e.g., EUR, AFR, EAS, AMR, SAS).

### --sum_stats:
The <ins>path</ins> to a GWAS summary statistics file. This file must be tab separated and contain at least these 4 following columns: SNP (containing SNP rsIDs), CHR (chromosome of the SNP), POS (position of SNP on chromosome), and P (p-value of association of SNP to disease). The following columns are not necessary but will be used by RegSCOUT if present: A1 (reference allele of SNP), A2 (alternative allele of SNP), MAF (minor allele frequency of SNP), and N (sample size). 

### --lead_snps: 
The <ins>path</ins> to a file containing the lead SNPs found in a GWAS study. This file must be tab separated if containing more than one column; however, only one column in this file is necessary for the pipeline to run: SNP (containing lead SNP rsIDs).

### --plink2_bin: 
This parameter is only required in the case that at least one of the A1, A2, or MAF columns is not present in the summary statistics provided to RegSCOUT. If this is the case, please provide to this parameter the <ins>path</ins> to the Plink2 executable downloaded from: https://www.cog-genomics.org/plink/2.0/. 

### --sample_num: 
This parameter can take an integer for the sample size of the GWAS (--sample_num 200000). However, if the N column (sample size) is present in the GWAS summary statistics, this column can be used instead of specifying an integer by setting this parameter as present (i.e., --sample_num present). 

### --fgwas_src:
The <ins>path</ins> to the fgwas executable after installation (e.g., --fgwas_src /home/fgwas/src/fgwas). 

### --CI_thr:
To SNPs have had their posterior probabilities of association (PPAs) assigned by fgwas, RegSCOUT finds the smallest set of SNPs in each locus whose cumulative PPAs add up to a certain threshold. This threshold is specified by this parameter by providing an integer. For example, if the smallest set of SNPs in each locus whose cumulative PPAs add up to 95% is desired then CI_thr can be set to 95 (i.e., --CI_thr 95).

## If mode is set to ATAC_obj (--mode ATAC_obj)

### Required parameters

### --output_dir: 
See description above.

### --jaspar_mtx:
In conducting TF analysis this pipeline uses the JASPAR TF binding profile database. If the user has downloaded position frequency matrices (PFMs) from JASPAR placed into a single file. This <ins>path</ins> to this file can be provided using this parameter. Alternatively, if the user would prefer not to upload such a file, this parameter can be set to none (i.e., --jaspar_mtx none). RegSCOUT in this case will obtain PFMs from JASPAR using the [JASPAR2024 library](https://bioconductor.org/packages/release/data/annotation/html/JASPAR2024.html). 

### --genome_built:
This parameter can be set to either hg19 or hg38 depending on the human genome build used in the GWAS data provided to RegSCOUT.

### --ci_gwas_dir:
If the finemap parameter was set to Y, this parameter should not be used as fine-mapping results will automatically be passed to RegSCOUT for downstream steps. If fine-mapping was not conducted using RegSCOUT, this parameter should be set to the <ins>path</ins> to the fine-mapping results. 

### --seurat_obj: 
The <ins>path</ins> to the Seurat object for the scATAC-seq data to be used in the integrative analysis. The provided Seurat object should have an Assay named "peaks" which holds a chromatin assay object created from the counts matrix of the scATAC-seq experiment. In addition, cell type labels should be provided as Idents in the Seurat object.   

### --peak_th:
This parameter is used to define a threshold, above which a open chromatin region is considered to be 'open' in a cell type. For example, if --peak_th is set to 0.1 (i.e., --peak_th 0.1), then greater than 10% of cells in a cell type must have nonzero scATAC-seq read counts at an open chromatin region, for an open chromatin region to be considered 'open' in that cell type.

### --prom_th_up:
This parameter is used to define the number of bases upstream a gene's transcription start site (TSS) that will be considered part of a gene's promoter. For example, if the user wants this value to be 2000 bases upstream the TSS they would set this parameter as 2000 (i.e., --prom_th_up 2000). 

### --prom_th_down:
This parameter is used to define the number of bases downstream a gene's TSS that will be considered part of the genes promoter. See example for --prom_th_up.

### --gencode_dir:
The <ins>path</ins> to the [GENCODE gene annotation](https://www.gencodegenes.org). This gene annotation should be a gff3 file.
