![logo](https://github.com/shooshtarilab/RegSCOUT/blob/main/images/RegSCOUT_logo.png)

# RegSCOUT (<ins>Reg</ins>ulatory <ins>S</ins>ingle <ins>C</ins>ell <ins>O</ins>mics for <ins>U</ins>nravelling <ins>T</ins>rait-Loci)
RegSCOUT is a computational pipeline that integrates disease-associated SNPs, identified through genome-wide association studies (GWAS), with scATAC-seq, Hi-C, eQTL, TF position frequency matrices (PFMs), scRNA-seq, and GENCODE gene annotation data. Through this comprehensive integrative approach, RegSCOUT prioritizes putative gene regulatory networks relevant to the pathogenesis of complex diseases. We have made RegSCOUT into an executable docker application for ease of use (<ins>development of docker application in progress</ins>); however, RegSCOUT can also be run in bash, example bash scripts can be found on the github page in the bash_scripts folder. While this README will provide a general overview of the pipeline and how it can be run, <ins>it is highly recommended that users read the more comprehensive user manual on this github before using RegSCOUT (RegSCOUT_User_Manual.docx)</ins>.

## Required R version and libraries:
RegSCOUT can be run on R versions 4.4.x and 4.5.x. It requires these libraries: tools, Seurat, Signac, GenomicRanges, genetics.binaRies, R.utils, atSNP, TFBSTools, JASPAR2024, stringr, circlize, ComplexHeatmap, R.utils, monocle3, cicero, Repitools, reshape2, dplyr, purrr, BSgenome.Hsapiens.UCSC.hg19, BSgenome.Hsapiens.UCSC.hg38, rtracklayer, data.table, writexl, tidyr, and tibble. <ins>Monocle3 must be installed prior to installing cicero</ins> (cicero analysis will not run properly if not). Cicero should be installed following instructions found in this link: https://cole-trapnell-lab.github.io/cicero-release/docs_m3/. 

## Note about genome builds:
When using RegSCOUT please ensure that all files provided to this pipeline are in the same genome build. E.g., PLINK bfiles (.bed, .bim, .fam), GWAS summary statistics, GWAS lead SNPs, GWAS locus region information, fine-mapped GWAS data, scATAC-seq datasets, Hi-C datasets, eQTL datasets, histone mark results datasets, and GENCODE gene information. RegSCOUT does not perform the lifting over of genome coordinates from one genome annotation version to another, this must be done prior to use of the pipeline. 

## Important Notes When Using RegSCOUT
1.	All directories specified in parameters passed to RegSCOUT should end with “/”. 
2.	Downstream stages of the pipeline will not work if the --mode parameter is not used. Please ensure you select a mode for RegSCOUT and provide scATAC-seq data to the pipeline.
3.	Please note that case (lower/upper case) matters in all aspects of this pipeline (parameters, column names for input files, etc.). Please keep this in mind when using parameters and creating input files, for example. 
4.	When providing numerical values to parameters, do not incorporate commas or spaces for thousands, e.g., use 100000 not 100,000.
5.	All chromosome numbers provided in files to RegSCOUT must be of the following naming convention 'chr' + 'number' (no space between). For example: chr8.
6.	Please see below explanation of mandatory RegSCOUT parameter (must always be specified):

| Parameter | Purpose and Potential Values | Default Value | 
| --------- | ---------------------------- | ------------- | 
| --output_dir | This parameter should specify the directory in which the user would like all outputs to be stored by RegSCOUT. This parameter should always be defined by the user. An ‘images’ folder will be created in this directory to store all figures generated by the pipeline. If this folder is already present in this directory, it will be overwritten. | No default, must be set by user. |

7.  Please see below explanations of important terms used in this README:

| Term | Definition | 
| ---- | ---------- | 
| Effect-SNPs | Fine-mapped SNPs that significantly influence the binding affinity of one or more TFs. |
| Risk-Mediating Peaks | Open chromatin regions in which one or more effect-SNPs colocalize (fall in the region). |

## Pipeline Overview
![Pipeline Overview](https://github.com/shooshtarilab/RegSCOUT/blob/main/images/Workflow_schematic.png)

Above is a general overview of the RegSCOUT workflow. Please note that certain parameters with defaults will not be explained in this README; intermediate outputs may not be explained as well. For a more comprehensive overview of each step, please see the user manual on this github. In the following sections of this README, descriptions of key parameters, inputs, and outputs, for each step, will be provided. Although the steps shown in the figure are numbered sequentially from 1-8, many are optional and dependent on what data is available to the user. 
- If --mode peak_table is requested, step 3 will not be run. 
- If the --finemap parameter is not specified, steps 1 & 2 will not run. 
- If the --hic_analysis parameter is not specified, step 5b will not run.
- If the --eqtl_analysis parameter is not specified, step 5c will not run.
- If the --histone_mark_analysis parameter is not specified, step 6 will not run.
- If the --tf_expr_analysis parameter is not specified, step 7 will not run.

## Fine-mapping Requested by User (--finemap Y; Steps 1 and 2)
To request that fine-mapping be conducted, the user must set the --finemap parameter to *Y*. RegSCOUT will perform preprocessing of GWAS summary statistics, create loci, and use fgwas (https://github.com/joepickrell/fgwas) to perform fine-mapping. These fine-mapping results will be integral to running subsequent steps of the RegSCOUT pipeline. 

### <ins>Parameters</ins>
| Parameter | Purpose and Potential Values | Default Value | 
| --------- | ---------------------------- | ------------- | 
| --finemap | Set this parameter to *Y* if fine-mapping should be conducted by RegSCOUT using fgwas. If fine-mapping is not desired (i.e., the user already has fine-mapping results), this parameter should not be used. | Analysis not conducted |
| --snp_ref_dir | Specifies the directory containing the bfiles used for PLINK preprocessing of GWAS data and LD analysis to identify loci. | No default, must be set by user |
| --population | Specifies the population/ancestry that the bfiles were created from (e.g., EUR, EAS). | No default, must be set by user |
| --sum_stats_dir | Specifies the path to the GWAS summary statistics. | No default, must be set by user |
| --lead_snps_dir | Specifies the path to GWAS lead SNP information. | No default, must be set by user |
| --plink2_dir | Specifies the path to the Plink2 executable. | No default, must be set by user |
| --sample_num | Sample size (total number of cases and controls) of the GWAS. If a sample size column (N) with sample size for each SNP is provided in the GWAS summary statistics, this parameter should be set to *present*. Alternatively, an integer value can be provided in this parameter to specify a constant overall sample size of the GWAS, this should only be provided if a sample size for each SNP is not readily available. | *present* |
| --fgwas_dir | Specifies the path to the fgwas executable. | No default, must be set by user |

### <ins>Input files</ins>

### GWAS Summary Statistics
The path to the summary statistics file should be specified using --sum_stats_dir. This file must be tab-separated and contain at least these 4 columns (note that RegSCOUT is case sensitive): ‘SNP’ (containing SNP rsIDs), ‘CHR’ (chromosome of the SNP), ‘POS’ (position of SNP on chromosome), and ‘P’ (p-value of association of SNP to disease). The following columns are not necessary but will be used by RegSCOUT if present: ‘A1’ (reference allele of SNP), ‘A2’ (alternative allele of SNP), ‘MAF’ (minor allele frequency of SNP), and ‘N’ (sample size). 

If the sample size column (‘N’) is present in the summary statistics, the --sample_num parameter should be set to *present*; however, if it is not, and the user is not able to find a sample size for each SNP, the overall sample size of the GWAS study should be provided using the --sample_num parameter (e.g., --sample_num 190000). Providing the ‘N’ column in the summary statistics is preferable and highly recommended.

If one of A1, A2, or MAF are missing from the GWAS summary statistics file, each of these columns will be filled in using Plink2 and its required bfiles (.bim, .bed, .fam). More information regarding these bfiles can be found below.

### GWAS Lead SNPs
The path to this file should be specified using --lead_snps_dir. This file must be tab-separated (.txt or .tsv) if containing more than one column; however, only one column in this file is necessary for the pipeline to run: ‘SNP’ (containing lead SNP rsIDs). 

### Plink bfiles
The path to these files is specified by the combination of the --snp_ref_dir and --population parameters. The file suffixes are not required in the path to these files, thus an example file path used by RegSCOUT is '/user/bin/EUR'. These are the .bim, .bed, and .fam files (e.g., EUR.bim, EUR.bed, and EUR.fam) required to run Plink2 to fill in reference/alternative allele as well as minor allele frequency information. These files will also be used to calculate LD values and group SNPs into loci. These files, in hg19, can be downloaded from here: https://mrcieu.github.io/gwasglue/ (see updated 1000 genomes reference panels). Users can also provide bfiles that they have generated, in hg38, or for the specific group they want to study.

### <ins>Output Files</ins>
| Output file | Description |
| ----------- | ----------- |
| final_gwas_data.txt | A tab-separated file organized by increasing locus number. This file is provided to fgwas for fine-mapping. There are 9 columns: ‘SNPID’ (the rsID of a SNP), ‘CHR’ (the chromosome the SNP is found on), ‘POS’ (the position of the SNP on the chromosome), ‘Z’ (the associated Z-score of the SNP), ‘F’ (the SNP's minor allele frequency), ‘N’ (sample size), ‘SEGNUMBER’ (locus number), ‘A1’ (reference allele of SNP), and ‘A2’ (alternative allele of SNP). |
| gwas_CI.txt | A tab-separated file which contains the results of filtering the results of fine-mapping. These are the credible interval SNPs. |

## Fine-mapping Already Conducted (--finemap not used)
In the case that fine-mapping results have already been obtained by the user, those results should be provided to RegSCOUT and the --finemap parameter should not be used. **The path to this file must be specified by the --ci_gwas_dir parameter.** The fine-mapping results file must contain credible interval SNPs, where, for each locus, the smallest number of SNPs whose cumulative PPAs add up to a certain threshold is provided (e.g., default 0.95; this threshold can be changed by the user). The file does not have a specific naming convention that it must follow but must be a tab-separated file (.txt or .tsv). This file must have at least 7 columns labelled: ‘id’ (SNP rsID), ‘pos’ (SNP base pair position), ‘chr’ (SNP chromosome), ‘chunk’ (locus number), ‘PPA’ (SNP PPA), ‘a1’ (SNP reference allele), and ‘a2’ (SNP alternative allele). Other columns that are optional but, if present, will be included in the final tables generated in step 8 (gene prioritization) of the pipeline are: 'lead_snp' (the lead SNP of a locus), 'locus_chr' (locus chromosome number e.g., chr8), 'locus_start' (start position of the locus), and 'locus_end' (end position of the locus). It is recommended that users provide this additional information about lead SNPs and loci. 

When creating this file please ensure that each SNP present in the file has <ins>just one reference and alternative allele</ins> listed in the ‘a1’ and ‘a2’ columns, respectively. Not doing so can cause errors when integrating SNP and TF motif data, an integral part of the RegSCOUT pipeline.

## Mode set to peak_table (--mode peak_table; Steps 4 and 5a)
The *peak_table* option should only be used if the user has already identified cell type-specific open chromatin regions and has already run cicero on their data to establish peak-to-peak links. If this is not the case, please see the section on --mode set to *ATAC_obj*. If the user has specified the *peak_table* option for RegSCOUT, this means that RegSCOUT will not be identifying cell type-specific open chromatin regions based on a certain threshold (e.g., a region must be open in 10% of cells of a cell type). RegSCOUT will also not be running cicero analysis to link open chromatin regions to each other as both these analyses must have already been conducted by the user. Instead, the user must provide both a peak-cell type table and cell type-specific co-accessibility tables (output of cicero) to RegSCOUT in the directory specified by the --output_dir parameter. More information on these tables and their formats can be found in the input files section.

### <ins>Parameters</ins>
| Parameter | Purpose and Potential Values | Default Value | 
| --------- | ---------------------------- | ------------- | 
| --mode | The user must indicate whether they have a scATAC-seq Seurat object or if they have a table of open chromatin regions as well as cicero results already generated. If the user is providing a Seurat object, this parameter should be set to *ATAC_obj*. If the user is providing a peak table and cicero results, this parameter should be set to *peak_table*. | No default, must be set by user |
| --ci_gwas_dir | Only use this parameter if fine-mapping <ins>was not</ins> conducted by RegSCOUT. Specifies the path to fine-mapping results. | If fine-mapping was not conducted by RegSCOUT: no default, must be set by user |
| --genome_build | The human reference genome that RegSCOUT should conduct analyses in, either *hg19* or *hg38*. | No default, must be set by user |
| --gencode_dir | Specifies the path to GENCODE gene annotation file downloaded from the GENCODE website (https://www.gencodegenes.org). | No default, must be set by user |
| --jaspar_mtx_dir | Specifies the path to a user provided file with JASPAR TF PFMs. If set to *none*, the JASPAR2024 R library will be used instead, to obtain TF PFMs. | *none* |

### <ins>Input files</ins>

### Peak-Cell Type Table
The path to this file need not be specified, RegSCOUT will look for this file in the directory defined by the --output_dir parameter. This file must be a tab-separated file (.tsv) named **cell_peak.tsv**. The table should have four columns: ‘chr’, ‘start’, ‘end’, and ‘cell_types’. Each row will represent one region of chromatin that was identified to be open. The ‘cell_types’ column will be a comma-separated list of all cell types in which the chromatin region on the same row was identified to be open. In the ‘cell_types’ column, there should be no spaces after each comma. 

Example peak-cell type table:
| chr | start | end | cell_types | 
| ---- | ----- | --- | -------------- | 
| chr1 | 123 | 1000 | plasma_cells,memory_b_cells | 
| chr11 | 2000 | 2600 | naive_CD8_T_cells,naive_CD4_T_cells | 
| chr18 | 1 | 678 | classical_monocytes,nonclassical_monocytes | 

### Cell Type-Specific Co-Accessibility Tables
The path to these files need not be specified as RegSCOUT will look for these files in the directory defined by the --output_dir parameter. These should be the output of cicero after filtering for a certain co-accessibility/significance threshold. RegSCOUT will not perform this filtering if --mode peak_table was selected. There will be one table for each of the cell types investigated. The names of each cell type-associated file should follow this naming convention: [cell type name].filtered_coaccessible_sites4s.txt (e.g., plasma_cells.filtered_coaccessible_sites4s.txt). These tables should be tab-separated and should have three columns named ‘Peak1’, ‘Peak2’, and ‘coaccess’ (co-accessibility value between peak1 and peak2 calculated by cicero). The Peak1 and Peak2 columns should contain underscore-separated locations of open chromatin regions (see example below).

Example cell type-specific co-accessibility table:
| Peak1 | Peak2 | coaccess | 
| ---- | ---- | ----- | 
| chr1_1000_1600 | chr1_2000_2600 | 0.13 |
| chr3_3000_4000 | chr3_5000_5500 | 0.27 |

### GENCODE Gene Annotation File
The user must provide a GENCODE gene annotation file downloaded from the GENCODE website: https://www.gencodegenes.org. The path to this file should be specified using the --gencode_dir parameter. Users should download the appropriate .gff3 file in the genome build that all other input files are in. The .gff3 file should be a comprehensive gene annotation file.

### JASPAR TF PFM File
In conducting SNP and TF motif integration, RegSCOUT uses the JASPAR TF binding motif database (https://jaspar.elixir.no). If the user would prefer not to create a comprehensive JASPAR TF PFMs file themselves, the jaspar_mtx_dir parameter can be set to *none* (i.e., --jaspar_mtx_dir none). This is the default value of this parameter. RegSCOUT, in this case, will obtain PFMs using the JASPAR2024 R library (https://bioconductor.org/packages/release/data/annotation/html/JASPAR2024.html). If users wish to provide their own file containing JASPAR TF PFMs, the path to this file should be specified using the --jaspar_mtx_dir parameter. To create this file, the user must download position frequency matrices (PFMs) from the JASPAR website and place them into a single file. A file which contains 755 human TF PFMs from the CORE JASPAR collection (combined_755_motifs_jaspar2024.txt) will be placed in the example_files folder on the github.

### <ins>Output files</ins>
| Output file | Description |
| ----------- | ----------- |
| Ci_effect_SNPs.txt | A tab-separated file containing all identified effect-SNPs, fine-mapped SNPs that significantly impacted the binding affinity of at least one TF, and their associated TFs. This file has 9 columns: ‘SNP’ (the rsID of a SNP), ‘CHR’ (the chromosome the SNP is found on), ‘Pos’ (the position of the SNP on the chromosome), ‘Locus’ (locus number), ‘TF’ (TF whose binding affinity was significantly impacted by the SNP), ‘Raw_P_value’ (p-value before FDR correction), ‘FDR_corr_P_value’ (p-value after FDR correction), 'log_like_ratio' (the log likelihood ratio comparing the impact of the alternative allele of a SNP on TF binding affinity compared to the reference allele), and 'ppa' (PPA of SNP). More information on log like ratio can be found in section 4.3 of the user manual. |
| cell_peak.svg | A visual representation of risk_regions_ppa.txt. This heatmap shows risk-mediating peaks on the y-axis and cell types on the x-axis. A heatmap square shaded in red indicates that a certain risk-mediating peak was found in a specific cell type whereas if a square is shaded in white, that risk-mediating peak was not found in that cell type. The bar on the right of this heatmap with shades of purple shows the sum of the PPAs of all effect-SNPs that fall into a risk-mediating peak (sumPPA). |
| cell_snp_tf.svg | A visual representation of risk_regions_ratio.txt specific to TF-SNP pairs. This heatmap shows TF-SNP pairs on the y-axis and cell types on the x-axis. A heatmap square shaded in blue indicates that a certain TF-SNP pair was found relevant to a specific cell type whereas if a square is shaded in white, that TF-SNP pair was not found relevant to that cell type. The bar on the right of this heatmap with shades of orange shows the log likelihood ratio value obtained for each TF-SNP pair. In this case, absolute valued log likelihood ratios are shown. |
| direct_rmp_gene_overlaps.txt | This is a table of all risk-mediating peak and gene promoter (as defined by GENCODE) overlaps identified, output as a tab-separated text file. There are 4 columns: 'RMP' (the risk-mediating peak), 'Promoter' (the promoter region defined using GENCODE), 'Gene' (the gene associated with that promoter), and 'Cell_Type' (the cell type in which the risk-mediating peak was identified). |
| cic_peak_interact_gene.txt | Table that shows links between RMPs and their target genes using peak-peak links identified by cicero, output as a tab-separated text file. This table has 6 columns: 'Peak1' (A risk-mediating peak), 'Peak2' (The peak that co-localized with a gene's promoter), 'coaccess' (the coaccessibility value calculated by cicero between Peak1 and Peak2), 'Promoter' (the promoter region defined using GENCODE), 'Gene' (the gene associated with that promoter), and 'Cell_Type' (the cell type in which this RMP-gene link was identified). |
| cell_gene_matrix.txt | A binary matrix that shows genes as columns and cell types as rows. A ‘1’ indicates that a gene was found relevant to a cell type according to risk-mediating peak and gene promoter linkage with cicero results, or by direct co-localization of a risk-mediating peak with a gene's promoter. A ‘0’ indicates that a gene was not found relevant to a cell type according to those criteria. |

## Mode set to ATAC_obj (--mode ATAC_obj; Steps 3, 4, and 5)
The *ATAC_obj* option should be used when the user has a scATAC-seq dataset that they would like RegSCOUT to use to identify risk-mediating open chromatin regions and run cicero analysis. More information on how this scATAC-seq dataset should be formatted can be found below. In this case, RegSCOUT will identify open chromatin regions in the scATAC-seq dataset based on a threshold that can be set by the user. RegSCOUT will also run cicero analysis, how it conducts cicero analysis can also be modified by the user, see section 5.1 of user manual for more details. 

### <ins>Parameters</ins>
Parameters listed below for --mode ATAC_obj are in addition to all parameters listed for --mode peak_table. In other words, the parameters relevant to --mode peak_table are a subset of those required for --mode ATAC_obj. 

| Parameter | Purpose and Potential Values | Default Value | 
| --------- | ---------------------------- | ------------- | 
| --seurat_obj_dir | Specifies the path to the scATAC-seq Seurat object. | No default, must be set by user |

### <ins>Input files</ins>

### scATAC-seq Seurat Object
The path to the scATAC-seq Seurat object should be provided using the --seurat_obj_dir parameter. The provided Seurat object should have an Assay named ‘peaks’ which holds a chromatin assay object created from the counts matrix of the scATAC-seq experiment. In addition, cell type labels should be provided as cell identities in the Seurat object (and accessible using the Idents() function of Seurat). This article provides more information on how to establish cell identities/class labels: https://satijalab.org/seurat/articles/essential_commands.html (under the ‘Identity class labels’ heading). The Seurat object can be provided to RegSCOUT as either an .RDS or .RData file. 

### GENCODE Gene Annotation File
See description of this file in 'Mode set to peak_table' section.

### JASPAR TF PFM File
See description of this file in 'Mode set to peak_table' section.

### <ins>Output files</ins>
Output files listed below are in addition to files output using --mode peak_table, please see 'Mode set to peak_table' section for more information on output files.

| Output file | Description |
| ----------- | ----------- |
| cell_peak.tsv | Table that provides genomic locations of peaks considered open in certain cell types. This table will have four columns: ‘chr’ (the chromosome the open chromatin region is found on), ‘start’ (start position of the region), ‘end’ (end position of the region), and ‘cell_types’ (all cell types in which the peak was considered open). |
| [cell type name].filtered_coaccessible_sites4s.txt | These files are the outputs of cicero analysis. They will have three columns: ‘Peak1’, ‘Peak2’, and ‘coaccess’ (the co-accessibility value calculated between Peak1 and Peak2). Only peak-peak links with co-accessibility greater than 0.05 will be found in these files (this threshold can be modified by the user, see section 5.1 of user manual for more details). One file will be output for each cell type in the scATAC-seq data. |

## Hi-C Analysis (--hic_analysis Y; Step 5b)
If analysis of Hi-C data, provided by the user, to assist in the prioritization of genes is desired, the user can set the --hic_analysis parameter to *Y*. 

### <ins>Parameters</ins>
| Parameter | Purpose and Potential Values | Default Value | 
| --------- | ---------------------------- | ------------- | 
| --hic_analysis | Set this parameter to *Y* if analysis of user-provided Hi-C data to assist in prioritization of genes is desired. If Hi-C analysis is not desired, do not use this parameter. | Analysis not conducted |
| --hic_instruct_dir | Specifies the path to user generated Hi-C analysis instructions file. | No default, must be set by user |

### <ins>Input files</ins>

### Hi-C Instructions File
For a thorough explanation of this instructions file, please see section 6.2 of the user manual on the github. An example of a Hi-C instructions file can be found in the instruction_files folder on the github.

### Hi-C Datasets
These datasets should be organized as tab-separated files (either as a .txt or .tsv) with a header. There are four different types of Hi-C files that could be provided to RegSCOUT, thorough descriptions of these four types and their required column names can be found in section 6.2 of the user manual. 

### <ins>Output files</ins>
| Output file | Description |
| ----------- | ----------- |
| all_hic_results.txt | The main output of Hi-C analysis. This is a table of all Hi-C interactions that linked risk-mediating peaks to genes. This table will have five columns: ‘cell’ (the scATAC-seq cell type in which the interaction was found), ‘gene’ (the gene name), ‘rmpRegion’ (the risk-mediating peak region), ‘promRegion’ (the Hi-C region that represents the gene’s promoter), ‘oeRegion’ (the Hi-C region interacting with the gene promoter that also colocalized with the risk-mediating peak region). |

## eQTL Analysis (--eqtl_analysis Y; Step 5c)
If analysis of eQTL data provided by the user, to assist in the prioritization of genes, is desired, the user can set the --eqtl_analysis parameter to *Y*.

### <ins>Parameters</ins>
| Parameter | Purpose and Potential Values | Default Value | 
| --------- | ---------------------------- | ------------- | 
| --eqtl_analysis | Set this parameter to *Y* if analysis of user provided eQTL data, to assist in prioritization of genes, is desired. If eQTL analysis is not desired, do not use this parameter. | Analysis not conducted |
| --eqtl_instruct_dir | Specifies the path to user-generated eQTL analysis instructions file. | Specifies the path to user-generated eQTL analysis instructions file. |

### <ins>Input files</ins>

### eQTL Instructions File
For a thorough explanation of this instructions file, please see section 7.2 of the user manual on the github. An example of an eQTL instructions file can be found in the instruction_files folder on the github.

### eQTL Datasets
RegSCOUT will not filter any of the eQTL datasets provided to it for significance, the user must conduct this filtering prior to using the pipeline. The pipeline also operates off gene names (e.g., TBXT) and not ENSEMBL gene IDs (e.g., ENSG00000164458). Gene IDs must be converted to gene names prior to using the pipeline. 

For datasets that are not being used for Tabix-based analysis, they should be organized as tab-separated files (either as a .txt or .tsv) with a header. These files should contain at least two columns: ‘snp’ (SNP rsIDs) and ‘gene’ (the gene with expression linked to the SNP). If only those two columns are provided, RegSCOUT will conduct eQTL analysis by matching SNPs based on rsIDs. Two other columns can be provided to avoid matching based on rsIDs if not desired: ‘chr’ (the chromosome the SNP is on) and ‘pos’ (the position of the SNP on the chromosome). If these two columns are provided alongside the ‘snp’ and ‘gene’ columns, RegSCOUT will conduct eQTL analysis by matching SNPs based on chromosome and position number. 

If Tabix-based analysis is desired, the user must provide their eQTL datasets as both a bgzipped file (.gz) and Tabix file (.tbi). If the user does not have these files, they can convert their tab-separated eQTL results file into these two files. More info on Tabix-based analysis and on converting to those two files can be found in section 7.2 of the user manual.

### <ins>Output files</ins>
| Output file | Description |
| ----------- | ----------- |
| all_eqtl_results.txt | The main output of eQTL analysis. This is a table of all eQTL results that colocalized with risk-mediating peak overlapping effect-SNPs, in a cell type-specific manner. This table will have four columns: ‘snp’ (SNP rsID), ‘rmp’ (the risk-mediating peak that the effect-SNP overlapped), ‘cell’ (the cell type that the SNP-gene link was established in), ‘gene’ (the name of the gene). |

## Histone Mark Analysis (--histone_mark_analysis Y; Step 6)
If analysis of histone mark data provided by the user, to investigate the functions of risk-mediating peaks, is desired, the user can set the –histone_mark_analysis parameter to *Y*. Histone mark analysis is conducted based on results generated by chromHMM (https://compbio.mit.edu/ChromHMM/).

### <ins>Parameters</ins>
| Parameter | Purpose and Potential Values | Default Value | 
| --------- | ---------------------------- | ------------- | 
| --histone_mark_analysis | Set this parameter to *Y* if analysis of user provided histone mark data, to investigate functions of risk-mediating peaks, is desired. If this is not desired, do not use this parameter. | Analysis not conducted |
| --hist_mark_instruct_dir | Specifies the path to user generated histone mark analysis instructions file. | No default, must be set by user |

### <ins>Input files</ins>

### Histone Mark Instructions File
For a thorough explanation of this instructions file, please see section 8.2 of the user manual on the github. An example of a histone mark instructions file can be found in the instruction_files folder on the github.

### Histone Mark Datasets
RegSCOUT will not do any preprocessing/filtering of these chromHMM results datasets that it is provided, any preprocessing/filtering must be conducted by the user prior to running the pipeline. These datasets must be cell type-specific, and each dataset should only contain chromHMM results for one cell type. These files should be tab-separated files (.tsv or .txt), with a header. Additionally, these files should contain at least these four columns: ‘chr’ (the chromosome of the region analyzed), ‘start’ (the start position of the region), ‘end’ (the end position of the region), and ‘state’ (the chromHMM label assigned to the region). 

### <ins>Output files</ins>
| Output file | Description |
| ----------- | ----------- |
| all_histone_mark_results.txt | The main output of histone mark analysis. This is a table of all risk-mediating peaks that colocalized with chromHMM results in a cell type-specific manner. This table will have four columns: ‘rmp’ (a risk-mediating peak), ‘state’ (the chromHMM labels associated with the peak), ‘version’ (the number of chromatin states chromHMM worked with e.g., 15-state or 18-state chromHMM), and ‘cell’ (the cell type that the function of this risk-mediating peak was established in). |

## TF Expression Analysis (--tf_expr_analysis atac, rna, or both; Step 7)
Disease-relevant TFs are first determined through identification of fine-mapped SNP-TF pairs in step 4 of RegSCOUT (note that the SNPs in these pairs are termed effect-SNPs). These SNP-TF pairs are then colocalized in step 5a with cell type-specific open chromatin regions – this is how cell type-specific TFs relevant to a disease are initially identified. If analysis of scATAC-seq data and/or scRNA-seq data provided by the user to contribute evidence toward the expression of TFs in the cell types they have been linked to in step 5a is desired, the --tf_expr_analysis parameter should be used. 

### <ins>Parameters</ins> 
| Parameter | Purpose and Potential Values | Default Value | 
| --------- | ---------------------------- | ------------- | 
| --tf_expr_analysis | Only use this parameter if TF expression analysis using scATAC-seq, scRNA-seq, or both is desired. Set to *atac* if TF expression analysis is desired using only scATAC-seq data that was provided as input for RegSCOUT earlier, for --mode peak_table or ATAC_obj. Set to *rna* if analysis is desired using only user-provided scRNA-seq data. Set to *both* if both scATAC-seq and scRNA-seq analyses are desired. E.g., --tf_expr_analysis both. | Analysis not conducted |
| --scrna_instruct_dir | Specifies the path to user-generated scRNA-seq analysis instructions file. Only required if scRNA-seq analysis has been requested, e.g., do not use this parameter if running --tf_expr_analysis atac. | No default, must be set by user |

### <ins>Input files</ins>

### scATAC-seq Seurat Object
Only one of a Seurat object or a peak-cell type table need to have been provided to RegSCOUT. This was provided to RegSCOUT if --mode was set to *ATAC_obj* and does not need to be provided again.

### Peak-Cell Type Table
Only one of a Seurat object or a peak-cell type table need to have been provided to RegSCOUT. This was provided to RegSCOUT if –mode was set to *peak_table* and does not need to be provided again.

### scRNA-seq Instructions File
For a thorough explanation of this instructions file, please see section 9.2 of the user manual on the github. An example of a scRNA-seq instructions file can be found in the instruction_files folder on the github.

### scRNA-seq Seurat Object
All scRNA-seq data should be provided in the form of a Seurat object. This data should be preprocessed before use in RegSCOUT. The provided Seurat objects should hold their gene-by-cell counts matrices in either the ‘counts’ or ‘data’ layer of the object. In addition, cell type labels should be provided as cell identities in the Seurat object (and accessible using the Idents() function of Seurat). This article provides more information on how to establish cell identities/class labels: https://satijalab.org/seurat/articles/essential_commands.html (under the ‘Identity class labels’ heading). The Seurat object can be provided to RegSCOUT as either an .RDS or .RData file, both will work.

### <ins>Output files</ins>
| Output file | Description |
| ----------- | ----------- |
| all_TF_expr_results.txt | The main output of TF expression analysis. This is a matrix with column names as cell types and row names as TFs. If --tf_expr_analysis was set to *both*, then the matrix can have values of ‘none’, ‘atac’, ‘rna’, or ‘both’, depending on what lines of evidence support the expression of a TF in a specific cell type. For example, if for the TF BACH2 in interneurons, the result was ‘rna’, then only scRNA-seq analysis supports the expression of that TF in that cell type. Whereas, if it was ‘both’ then both scATAC-seq and scRNA-seq analysis support its expression. If it was ‘none’ then neither scATAC-seq or scRNA-seq support the expression of that TF in the cell type. If --tf_expr_analysis was set to *atac*, then only the values of ‘none’ or ‘atac’ will be present in the matrix. If --tf_expr_analysis was set to *rna*, then only the values of ‘none’ or ‘rna’ will be present in the matrix. |
| cell_tf.svg | This heatmap combines results of TF-SNP pair colocalization with cell type-specific open chromatin regions, and results of cell type-specific TF expression analysis. Cell types can be found on the x-axis and TFs can be found on the y-axis. This heatmap will have up to 5 different categories in its legend depending on the type of TF expression analysis requested by the user (i.e., atac, rna, or both). The labels on the legend include ‘no evidence’ (when no evidence for the relevance of the TF in the cell type is found), ‘snp impact’ (when the TF’s binding affinity is impacted by a SNP and that SNP-TF pair colocalizes to an open chromatin region in that cell type), ‘snp impact + atac’ (includes the criteria of ‘snp impact’ + scATAC-seq analysis supports the expression of the TF in that cell type), ‘snp impact + rna’ (includes the criteria of ‘snp impact’ + scRNA-seq analysis supports the expression of the TF in that cell type), and ‘snp impact + atac + rna’ (includes the criteria of ‘snp impact’ and both scRNA-seq and scATAC-seq analysis support the expression of the TF in that cell type). |

## Gene Prioritization (Step 8)
This step will always run as long as the --mode parameter is used; no other parameter must be specified for this step to run. This step will perform gene prioritization based on criteria that can be set by the user, but defaults are available. More information on the criteria used can be found in section 10 of the user manual.

### <ins>Parameters</ins>
See section 10 of user manual for explanations of gene score, gene sum PPA, and TF score. Set any of these parameters to 0 if prioritization using that parameter is not desired (note: gene score will always have a value of at least 1). By default, RegSCOUT does not prioritize using TF score.

| Parameter | Purpose and Potential Values | Default Value | 
| --------- | ---------------------------- | ------------- | 
| --gene_score_th | The threshold for gene score (lines of evidence supporting a gene) that a gene should obtain a score <ins>greater than or equal to</ins> in order to be prioritized. | 2 |
| --gene_sum_ppa_th | Genes will only be prioritized if their sum ppa value is <ins>greater than or equal to</ins> this threshold. | 0.05 |
| --tf_score_th | A SNP-TF pair to gene link will only be prioritized if the TF attains a score <ins>greater than or equal to</ins> this threshold. | 0 |

### <ins>Output files</ins>
| Output file | Description |
| ----------- | ----------- |
| prioritized_table_condensed.tsv/xlsx | This is the most condensed table output by RegSCOUT, which only contains results after gene prioritization. This is a summary table for all prioritized results. An explanation of columns in this table can be found in section 10.2 of the user manual. |
| cell_gene.svg | This is a heatmap that shows the genes that were prioritized by RegSCOUT and the cell types that they were prioritized in. The x-axis consists of cell types and prioritized gene names are located on the y-axis. The colour of each heatmap square corresponds to the gene score obtained by that gene in that cell type. |

## Using Bash Scripts
RegSCOUT is available as a docker application for ease of use; however, it can also be run using bash scripts. Example bash scripts can be found on the github repository, in the bash_scripts folder. Section 11 of the user manual offers further explanation for how RegSCOUT can be run in bash. 
