![logo](https://github.com/shooshtarilab/RegSCOUT/blob/main/images/RegSCOUT_logo.png)

# RegSCOUT (<ins>Reg</ins>ulatory <ins>S</ins>ingle <ins>C</ins>ell <ins>O</ins>mics for <ins>U</ins>nravelling <ins>T</ins>rait Loci - Updates in Progress)
This is the repository for the RegSCOUT pipeline. This pipeline can be used as a stand-alone command-line application which integrates disease-associated SNPs, identified through genome-wide association studies (GWAS), with scATAC-seq, Hi-C, and eQTL data. Through this integrative analysis, RegSCOUT prioritizes putative gene regulatory networks relevant to complex disease pathogenesis. Example bash scripts for use of RegSCOUT can be found in the bash_scripts folder on this repository. 

This pipeline can handle both a scATAC-seq data as a Seurat object or processed results of analyzing scATAC-seq data as a peak and cell type table along with peak-peak interaction tables (generated by [cicero](https://bioconductor.org/packages/release/bioc/html/cicero.html)). See the --mode peak_table section for more information about these two tables. 

## Required R version and libraries:
RegSCOUT was created using R version 4.4.0 and uses these libraries: Seurat, Signac, GenomicRanges, genetics.binaRies, R.utils, atSNP, TFBSTools, stringr, circlize, ComplexHeatmap, xlsx, R.utils, ape, monocle3, cicero, Repitools, reshape2, dplyr, BSgenome.Hsapiens.UCSC.hg19, BSgenome.Hsapiens.UCSC.hg38, ggplot2, readxl, tidyr, tibble. Monocle3 must be installed prior to installing cicero. Cicero should be installed following instructions found in this link: https://cole-trapnell-lab.github.io/cicero-release/docs_m3/. 

## Note about genome builds:
When using RegSCOUT please ensure that all files provided to this pipeline are in the same genome build. E.g., PLINK bfiles (.bed, .bim, .fam), GWAS summary statistics, GWAS lead SNPs, GWAS locus region information, fine-mapped GWAS data, scATAC-seq datasets, Hi-C datasets, eQTL datasets, histone mark results datasets, and GENCODE gene information. 

## Pipeline Overview
![Pipeline Overview](https://github.com/shooshtarilab/RegSCOUT/blob/main/images/Workflow_schematic.png)

Above is a general overview of the RegSCOUT workflow. Each step will be explained below with a description of inputs and outputs. A table of all parameters included in RegSCOUT including which steps the parameter is relevant to, potential values, default value, and general purpose of the parameter can also be found below.

## Table of All Parameters
| Parameter | Purpose and Potential Values | Default Value | Relevant Steps |
| --------------------- | ----------------------------------------------------- | ---------------- | ---------------------- |
| --snp_ref_dir | Specifies the directory containing the bfiles used for PLINK preprocessing of GWAS data and LD analysis. | No default, must be set by user. | 1 |
| --population | Specifies the population/ancestry that the bfiles were created from (e.g., EUR, EAS). | No default, must be set by user. | 1 |
| --sum_stats_dir | Specifies the path to the GWAS summary statistics | No default, must be set by user | 1 |
| --lead_snps_dir | Specifies the path to GWAS lead SNP information | No default, must be set by user | 1 |
| --plink2_dir | Specifies the path to the Plink2 executable | No default, must be set by user | 1 |
| --sample_num | Sample size (total number of cases and controls) of the GWAS. If a sample size column (N) with sample size for each SNP is provided in the GWAS summary statistics, this parameter should be set to *present*. Alternatively, an integer value can be provided in this parameter to specify the overall sample size of the GWAS, this should only be provided if a sample size for each SNP is not readily available. | *present* | 1 |
| --locus_region | The # of bps to add on both sides of lead SNP positions to define locus regions. | 1,000,000 | 1 |
| --ld_th | SNPs are not included in a locus if they obtain an abs(LD value) (absolute value), with the lead SNP, less than or equal to this threshold. | 0.25 | 1 |
| --fgwas_dir | Specifies the path to the fgwas executable | No default, must be set by user | 2 |
| --ci_th | For each locus, RegSCOUT filters for the smallest group of SNPs whose cumulative posterior probabilities of association (PPAs) add up to this threshold. | 0.95 | 2 |
| --ci_ppa_th | All CI SNPs must have a PPA greater than this threshold. | 0.01 | 2 |
| --finemap | Set this parameter to *Y* if fine-mapping should be conducted by RegSCOUT using fgwas. If fine-mapping is not desired (i.e., the user already has fine-mapping results), this parameter should not be used. | Not conducted | 1, 2 |
| --seurat_obj_dir | Specifies the path to the scATAC-seq Seurat object. | No default, must be set by user | 3 |
| --cell_count_th | Cell types with less than this number of cells in the scATAC-seq dataset will be removed. | 3 | 3 |
| --coaccess_th | The co-accessibility value between two peaks must be greater than this threshold to be output in cicero results. | 0.05 | 3 |
| --cic_genomic_window | Size of the regions on either side of an open chromatin region that cicero will look for peak to peak links in. | 2,000,000 | 3 |
| --jaspar_mtx_dir | Specifies the path to file with JASPAR TF PFMs. If set to *none*, the JASPAR2024 R library will be used to obtain TF PFMs. | *none* | 4 |
| --ci_gwas_dir | Only use this parameter if fine-mapping **was not** conducted by RegSCOUT. Specifies the path to fine-mapping results. | If fine-mapping was not conducted by RegSCOUT: no default, must be set by user | 4 |
| --genome_build | The reference genome that RegSCOUT should conduct analyses in, either hg19 or hg38. | No default, must be set by user | 3, 4 |
| --peak_th | The threshold above which a peak is considered to be open in a cell type. E.g., a peak can be considered open if it is accessible in 1/10 or 10% of cells of a cell type. | 0.1 | 3, 4 |
| --prom_th_up | Number of bps to add upstream a TSS to define the gene promoter. | 2,000 | 5 |
| --prom_th_down | Number of bps to add downstream a TSS to define the gene promoter. | 2,000 | 5 |
| --mode | The user must indicate whether they have a scATAC-seq Seurat object or if they have a table of open chromatin regions as well as cicero results already generated. If the user is providing a seurat object, this parameter should be set to *ATAC_obj*. If the user is providing a peak table and cicero results, this parameter should be set to *peak_table*. | No default, must be set by user | 3, 4, 5 |
| --hic_analysis | Set this parameter to *Y* if gene regulatory analysis using Hi-C analysis is desired. If Hi-C analysis is not desired, do not use this parameter | Analysis not conducted | 6 |
| --hic_instruct_dir | Specifies the path to user generated Hi-C analysis instructions. | No default, must be set by user | 6 |
| --eqtl_analysis | Set this parameter to *Y* if gene regulatory analysis using eQTL analysis is desired. | Analysis not conducted | 7 |
| --eqtl_instruct_dir | Specifies the path to user generated eQTL analysis instructions. | No default, must be set by user | 7 |
| --histone_mark_analysis | Set this parameter to *Y* if analysis of risk-mediating peaks using user-provided histone mark data is desired | Analysis not conducted | 8 |
| --hist_mark_instruct_dir | Specifies the path to user generated histone mark analysis instructions. | No default, must be set by user | 8 |
| --tf_expr_analysis | Only use this parameter if TF expression analysis using scATAC-seq, scRNA-seq, or both is desired. Set to *atac* if TF expression analysis is desired using scATAC-seq data that was provided as input earlier for --mode *peak_table* or *ATAC_obj*. Set to *rna* if analysis is desired using user-provided scRNA-seq data. Set to *both* if both scATAC-seq and scRNA-seq analyses are desired. | No conducted | 9 |
| --scrna_instruct_dir | Specifies the path to user generated scRNA-seq analysis instructions spreadsheet. | No default, must be set by user | 9 |
| --tf_rna_quantile_th | The quantile for percent expression of genes in a cell type, above which a TF will be considered expressed in that cell type. | 0.25 | 9 |
| --gencode_dir | Specifies the path to GENCODE gene information. | No default, must be set by user | 5, 6, 9 |
| --tf_score_th | Genes will only be prioritized if one of their associated TFs attains a score higher than this threshold. | -1 | 10 |
| --gene_sum_ppa_th | Genes will only be prioritized if their sum ppa value is greater than this threshold. | 0.05 | 10 |
| --gene_score_th | Genes will only be prioritized if their gene score is greater than this threshold. | 1 | 10 |
| --output_dir | The directory in which all RegSCOUT output files will be stored. | No default, must be set by user | 1 through 10 |

## Fine-mapping (Steps 1 and 2)

## Main Parameters:
### --mode: 
This parameter should be used to choose the mode of operation of the pipeline based on input type. If a scATAC-seq Seurat object is provided, this option should be set to ATAC_obj (i.e., --mode ATAC_obj). If a peak and cell type table and peak-peak interaction table are used this mode should be set to peak_table (i.e., --mode peak_table).

### --finemap:
This parameter should be used only when the user wants fine-mapping to be conducted within the RegSCOUT workflow using [fgwas](https://github.com/joepickrell/fgwas). If this is the case --finemap should be set to Y (i.e., --finemap Y). If this is not the case, this parameter should not be used. 

### --output_dir: 
This parameter should be used for setting the working directory (the directory address should end with "/"). **This parameter should always be set when using RegSCOUT regardless of the value of --finemap or --mode**. All the output and intermediate files of the pipeline are generated in this directory. If running the pipeline with --mode peak_table, the peak-by-cell and peak-peak interaction tables of the pipeline should be provided in this directory.

*All directories specified in parameters should end with "/"*

*Please note that case (lower/upper case) matters in all aspects of this pipeline (parameters, column names for input files, etc.)*

## If fine-mapping must be conducted (--finemap Y)

### <ins>Required parameters</ins>

### --output_dir: 
See description above.

### --SNP_ref:
This parameter specifies the <ins>directory</ins> containing LD reference files from the 1000 Genomes project (bed, bim, fam files). These files could be downloaded from https://mrcieu.github.io/gwasglue/ (see Reference datasets subheading).

### --Population:
This is a three letter identifier for the LD reference population (e.g., EUR, AFR, EAS, AMR, SAS).

### --sum_stats:
The <ins>path</ins> to a GWAS summary statistics file. This file must be tab separated and contain at least these 4 following columns: SNP (containing SNP rsIDs), CHR (chromosome of the SNP), POS (position of SNP on chromosome), and P (p-value of association of SNP to disease). The following columns are not necessary but will be used by RegSCOUT if present: A1 (reference allele of SNP), A2 (alternative allele of SNP), MAF (minor allele frequency of SNP), and N (sample size). 

### --lead_snps: 
The <ins>path</ins> to a file containing the lead SNPs found in a GWAS study. This file must be tab separated if containing more than one column; however, only one column in this file is necessary for the pipeline to run: SNP (containing lead SNP rsIDs).

### --plink2_bin: 
This parameter is only required in the case that at least one of the A1, A2, or MAF columns is not present in the summary statistics provided to RegSCOUT. If this is the case, please provide to this parameter the <ins>path</ins> to the Plink2 executable downloaded from: https://www.cog-genomics.org/plink/2.0/. 

### --sample_num: 
This parameter can take an integer for the sample size of the GWAS (--sample_num 200000). However, if the N column (sample size) is present in the GWAS summary statistics, this column can be used instead of specifying an integer by setting this parameter as present (i.e., --sample_num present). 

### --fgwas_src:
The <ins>path</ins> to the fgwas executable after installation (e.g., --fgwas_src /home/fgwas/src/fgwas). 

### --CI_thr:
To SNPs have had their posterior probabilities of association (PPAs) assigned by fgwas, RegSCOUT finds the smallest set of SNPs in each locus whose cumulative PPAs add up to a certain threshold. This threshold is specified by this parameter by providing an integer. For example, if the smallest set of SNPs in each locus whose cumulative PPAs add up to 95% is desired then CI_thr can be set to 95 (i.e., --CI_thr 95).

### <ins>Output files</ins>

### Plink2.afreq
A file output by Plink2 which provides SNP information including reference and alternative alleles and minor allele frequency. Will not be output if A1, A2, and MAF files are present in GWAS summary statistics. 

### new_lead_snps.txt
If the A1, A2, or MAF column in missing in the GWAS summary statistics, this file will contain only those lead SNPs that were found in Plink2, which subsequent analysis will be performed on. If the three columns were present this new_lead_snps.txt file should not be different from the lead SNP file provided initially to RegSCOUT.

### loci_info.txt
File containing information on each locus region created by RegSCOUT. The columns of this file are SEGNUM (the locus number), lead_snp (the lead SNP of the locus), locus_chr (the chromosome the locus is found on), locus_start (start position of the locus), and locus_end (end position of the locus).

### final_gwas_data.txt
A tab-separated file organized by increasing locus number. This file is provided to fgwas for fine-mapping. There are 9 columns: SNPID (the rsID of a SNP), CHR (the chromosome the SNP is found on), POS (the position of the SNP on the chromosome), Z (the associated Z-score of the SNP), F (the SNP's minor allele frequency), N (sample size), SEGNUMBER (locus number), A1 (reference allele of SNP), A2 (alternative allele of SNP). 

### files starting with CI (e.g., CI.bfs.gz)
These are output files produced by fgwas as it assigns PPAs to SNPs. CI.bfs.gz contains SNPs and their associated PPA values. 

### gwas_CI.txt
A tab-separated file which contains the results of fine-mapping and filtering for CI_thr. SNPs on sex chromosomes and with PPA values <= 0.01 have also been filtered out. These are the credible interval SNPs.

## If mode is set to ATAC_obj (--mode ATAC_obj)

### <ins>Required parameters</ins>

### --output_dir: 
See description above.

### --jaspar_mtx:
In conducting TF analysis this pipeline uses the JASPAR TF binding profile database. If the user has downloaded position frequency matrices (PFMs) from JASPAR placed into a single file. This <ins>path</ins> to this file can be provided using this parameter. Alternatively, if the user would prefer not to upload such a file, this parameter can be set to none (i.e., --jaspar_mtx none). RegSCOUT in this case will obtain PFMs from JASPAR using the [JASPAR2024 library](https://bioconductor.org/packages/release/data/annotation/html/JASPAR2024.html). 

### --genome_built:
This parameter can be set to either hg19 or hg38 depending on the human genome build used in the GWAS data provided to RegSCOUT.

### --ci_gwas_dir:
If the finemap parameter was set to Y, this parameter should not be used as fine-mapping results will automatically be passed to RegSCOUT for downstream steps. If fine-mapping was not conducted using RegSCOUT, this parameter should be set to the <ins>path</ins> to the fine-mapping results. 

### --seurat_obj: 
The <ins>path</ins> to the Seurat object for the scATAC-seq data to be used in the integrative analysis. The provided Seurat object should have an Assay named "peaks" which holds a chromatin assay object created from the counts matrix of the scATAC-seq experiment. In addition, cell type labels should be provided as Idents in the Seurat object. The seurat object should be saved as an RData file. 

### --peak_th:
This parameter is used to define a threshold, above which a open chromatin region is considered to be 'open' in a cell type. For example, if --peak_th is set to 0.1 (i.e., --peak_th 0.1), then greater than 10% of cells in a cell type must have nonzero scATAC-seq read counts at an open chromatin region, for an open chromatin region to be considered 'open' in that cell type.

### --prom_th_up:
This parameter is used to define the number of bases upstream a gene's transcription start site (TSS) that will be considered part of a gene's promoter. For example, if the user wants this value to be 2000 bases upstream the TSS they would set this parameter as 2000 (i.e., --prom_th_up 2000). 

### --prom_th_down:
This parameter is used to define the number of bases downstream a gene's TSS that will be considered part of the genes promoter. See example for --prom_th_up.

### --gencode_dir:
The <ins>path</ins> to the [GENCODE gene annotation](https://www.gencodegenes.org). This gene annotation should be a gff3 file.

### <ins>Output files</ins>

### atSNP_10runs_results.RDS
An RDS file containing the results of 10 runs of the ComputePValues function of the atSNP R package which identifies the impact of SNPs on TF binding affinities. 

### Ci_effect_SNPs.txt
A tab-separated file containing all identified effect-SNPs and their associated TFs. Effect-SNPs are SNPs which significantly affected the binding affinity of at least one TF. This file has 9 columns: SNP (the rsID of a SNP), CHR (the chromosome the SNP is found on), Pos (the position of the SNP on the chromosome), Locus (locus number), TF (TF whose binding affinity was significantly impacted by SNP), Raw_P_value (p-value before FDR correction), FDR_corr_P_value (p-value after FDR correction), log_like_ratio (the log likelihood ratio illustrating the impact of the alternative allele of a SNP on TF binding compared to the reference allele), and ppa (PPA of SNP).

### cell_peak.xlsx 
Table that provides genomic locations of peaks considered open in certain cell types (i.e., those that met the CI_thr). This table will have four columns: chr (the chromosome the open chromatin region is found on), start (start position of open chromatin region), end (end position of open chromatin region), and cell sub-types (all cell types in which a peak was considered open).

### [cell type name].filtered_coaccessible_sites4s.txt
These files are the results of cicero analysis. They will have three columns: Peak1, Peak2, and coaccess (the co-accessibility value calculated between Peak1 and Peak2). Only peak-peak interactions with co-accessibility > 0.05 will be found in these files. One file will be output for each cell type in the scATAC-seq data. 

### risk_regions_ppa.xlsx
Table that provides information on risk-mediating peaks (RMPs) found by RegSCOUT. RMPs are those open chromatin regions that contain at least one effect-SNP. It has three columns: region (the RMP), cell (the cell types in which the region was found to be an RMP), and sumPPA (the sum of the PPAs of all effect-SNPs that fall in that RMP).

### peak_cluster_matrix.txt
A binary matrix that contains cell types as columns and RMPs as rows. A '1' indicates that that a RMP was found in a specified cell type whereas a '0' means it was not. 

### cell_peak.png
A visual representation of risk_regions_ppa.xlsx. This heatmap shows RMPs on the y-axis and cell types on the x-axis. A heatmap cell shaded in red indicates that a certain RMP was found in a certain cell type whereas if a cell is shaded in white, that RMP was not found in that cell type. The bar on the right of this heatmap with shades of blue shows the sumPPA value obtained for each RMP.

### risk_regions_ratio.xlsx
Table providing information on RMPs and their associated effect-SNPs. This table has 4 columns: TFSNP (a TF-SNP pair identified by atSNP), region (an RMP), cell (the cell type in which this effect SNP was found to fall in this RMP), and log_lik_ratio (the log likelihood ratio describing an effect-SNPs effect on the TF's binding affinity).

### TF_cluster_matrix.txt
A binary matrix that contains cell types as columns and TF-SNP pairs as rows. If a TF-SNP pair was found relevant to a certain cell type it would obtain a value of '1', otherwise it would obtain a value of '0'.

### cell_tf.png
A visual representation of risk_regions_ratio specific to TF-SNP pairs. This heatmap shows TF-SNP pairs on the y-axis and cell types on the x-axis. A heatmap cell shaded in red indicates that a certain TF-SNP pair was found relevant to a certain cell type whereas if a cell is shaded in white, that TF-SNP pair was not found relevant to that cell type. The bar on the right of this heatmap with shades of blue shows the log likelihood ratio value obtained for each TF-SNP pair.

### Aff_peak_interact_gene.csv
Table that shows the links between RMPs and their target genes. This table has 6 columns: Peak1 (An RMP), Peak2 (the co-accessible peak associated with the RMP), coaccess (the co-accessibility value between those two peaks), Promoter (the promoter region of the affected gene defined by --prom_th_up and --prom_th_down parameters), Gene (the id of affected gene), and Cell_Type (the cell type in which this putative RMP-gene interaction was detected). Please note that the rows with identical Peak1 and Peak2 represent the RMPs which directly overlap the promoters of their target genes.

### cell_gene.csv
A binary matrix that shows genes as columns and cell types as rows. A '1' indicates that a gene was prioritized in a cell type according to cicero (co-accessibility value > 0.05 between an RMP and the gene's promoter), whereas a value of '0' indicates the gene was not prioritized in that cell type. 

### cell_gene.png
A visual representation of cell_gene.csv. This heatmap shows cell types on the y-axis and genes on the x-axis. A heatmap cell shaded in red indicates that a certain gene was prioritized to a certain cell type whereas if a cell is shaded in white, that gene was not found relevant to that cell type according to cicero.

### Gene_promoter_matrix.txt
A binary matrix that shows cell types as columns and RMP-gene pairs as rows. This matrix is specific to those genes whose promoters were overlapped by an RMP. A '1' indicates that that RMP-gene pair was identified in that cell type, whereas a '0' indicates that RMP-gene pair was not found relevant to that cell type. 

### cell_gene_promoter.png
A visual representation of Gene_promoter_matrix.txt. This heatmap is specific to those genes whose promoters were overlapped by an RMP. This heatmap shows RMP-gene pairs on the y-axis and cell types on the x-axis. A heatmap cell shaded in red indicates that a certain RMP-gene pair was found relevant to a certain cell type whereas if a cell is shaded in white, that pair was not found relevant to that cell type.

### Gene_enhancer_matrix.txt
A binary matrix that shows cell types as columns and RMP-gene pairs as rows. This matrix is specific to the case where cicero identified a relationship between an RMP and a gene's promoter where the RMP did not overlap the gene's promoter. A '1' indicates that that RMP-gene pair was identified in that cell type, whereas a '0' indicates that RMP-gene pair was not found relevant to that cell type. 

### cell_gene_enhancer.png
A visual representation of Gene_enhancer_matrix.txt. This heatmap is specific to those genes whose promoters were found to be co-accessible with an RMP but the RMP did not directly overlap the promoter. This heatmap shows RMP-gene pairs on the y-axis and cell types on the x-axis. A heatmap cell shaded in red indicates that a certain RMP-gene pair was found relevant to a certain cell type whereas if a cell is shaded in white, that pair was not found relevant to that cell type.

## If mode is set to peak_table (--mode peak_table)

### <ins>Required parameters</ins>
Parameters required for --mode peak_table are nearly the same as those required for --mode ATAC_obj; the difference is that the --seurat_obj and --peak_th parameters are no longer necessary. The peak and cell type table and peak-peak interaction tables should be placed in the directory specified by --output_dir. 

### <ins>The peak and cell type table, and peak-peak interaction tables</ins>
This peak and cell type table should be named cell_peak.xlsx and should have three columns named "chr", "start", "end" (chromosome, start and end of the open chromatin regions), and a column named “cell sub-types” which represents the cell types in which a peak is considered open. Multiple cells should be comma-separated in this column. There should be one peak-peak interaction table for each cell type to be studied. The names of their associated files should be [cell type name].filtered_coaccessible_sites4s.txt. These tables should be tab-separated and should have three columns named "Peak1", "Peak2", and "coaccess" (coaccessibility value between peak1 and peak2 calculated by cicero). The Peak1 and Peak2 columns should contain underscore-separated locations of open chromatin regions (e.g., chr1_1000_1600). Examples of a peak and cell type table and of a peak-peak interaction table are shown below. 

Example peak and cell type table:
| chr | start | end | cell sub-types | 
| ---- | ----- | --- | -------------- | 
| chr1 | 123 | 1000 | plasma_cells,memory_b_cells | 
| chr11 | 2000 | 2600 | naive_CD8_T_cells,naive_CD4_T_cells | 
| chr18 | 1 | 678 | classical_monocytes,intermediate_monocytes,nonclassical_monocytes | 

Example peak-peak interaction table:
| Peak1 | Peak2 | coaccess | 
| ---- | ---- | ----- | 
| chr1_1000_1600 | chr1_2000_2600 | 0.13 |
| chr3_3000_4000 | chr3_5000_5500 | 0.27 |

### <ins>Output files</ins>
The output for this mode would be the same as those listed in --mode ATAC_obj; the difference is that cell_peak.xlsx and [cell type name].filtered_coaccessible_sites4s.txt are not output as they are required as inputs when --mode is set to peak_table. 
